// @ark-ui/react/tooltip/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const tooltip = require('./tooltip.cjs');
const tooltipArrow = require('./tooltip-arrow.cjs');
const tooltipArrowTip = require('./tooltip-arrow-tip.cjs');
const tooltipContent = require('./tooltip-content.cjs');
const tooltipContext = require('./tooltip-context.cjs');
const tooltipPositioner = require('./tooltip-positioner.cjs');
const tooltipTrigger = require('./tooltip-trigger.cjs');

const Tooltip = Object.assign(tooltip.Tooltip, {
  Root: tooltip.Tooltip,
  Arrow: tooltipArrow.TooltipArrow,
  ArrowTip: tooltipArrowTip.TooltipArrowTip,
  Content: tooltipContent.TooltipContent,
  Positioner: tooltipPositioner.TooltipPositioner,
  Trigger: tooltipTrigger.TooltipTrigger
});

exports.TooltipArrow = tooltipArrow.TooltipArrow;
exports.TooltipArrowTip = tooltipArrowTip.TooltipArrowTip;
exports.TooltipContent = tooltipContent.TooltipContent;
exports.useTooltipContext = tooltipContext.useTooltipContext;
exports.TooltipPositioner = tooltipPositioner.TooltipPositioner;
exports.TooltipTrigger = tooltipTrigger.TooltipTrigger;
exports.Tooltip = Tooltip;
