// @ark-ui/react/menu/index.mjs
'use client';
import { Menu as Menu$1 } from './menu.mjs';
import { MenuArrow } from './menu-arrow.mjs';
import { MenuArrowTip } from './menu-arrow-tip.mjs';
import { MenuContent } from './menu-content.mjs';
export { useMenuContext } from './menu-context.mjs';
import { MenuContextTrigger } from './menu-context-trigger.mjs';
import { MenuItem } from './menu-item.mjs';
import { MenuItemGroup } from './menu-item-group.mjs';
import { MenuItemGroupLabel } from './menu-item-group-label.mjs';
import { MenuOptionItem } from './menu-option-item.mjs';
import { MenuPositioner } from './menu-positioner.mjs';
import { MenuSeparator } from './menu-separator.mjs';
import { MenuTrigger } from './menu-trigger.mjs';
import { MenuTriggerItem } from './menu-trigger-item.mjs';

const Menu = Object.assign(Menu$1, {
  Root: Menu$1,
  Arrow: MenuArrow,
  ArrowTip: MenuArrowTip,
  Content: MenuContent,
  ContextTrigger: MenuContextTrigger,
  Item: MenuItem,
  ItemGroup: MenuItemGroup,
  ItemGroupLabel: MenuItemGroupLabel,
  OptionItem: MenuOptionItem,
  Positioner: MenuPositioner,
  Separator: MenuSeparator,
  Trigger: MenuTrigger,
  TriggerItem: MenuTriggerItem
});

export { Menu, MenuArrow, MenuArrowTip, MenuContent, MenuContextTrigger, MenuItem, MenuItemGroup, MenuItemGroupLabel, MenuOptionItem, MenuPositioner, MenuSeparator, MenuTrigger, MenuTriggerItem };
