// @ark-ui/react/menu/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const menu = require('./menu.cjs');
const menuArrow = require('./menu-arrow.cjs');
const menuArrowTip = require('./menu-arrow-tip.cjs');
const menuContent = require('./menu-content.cjs');
const menuContext = require('./menu-context.cjs');
const menuContextTrigger = require('./menu-context-trigger.cjs');
const menuItem = require('./menu-item.cjs');
const menuItemGroup = require('./menu-item-group.cjs');
const menuItemGroupLabel = require('./menu-item-group-label.cjs');
const menuOptionItem = require('./menu-option-item.cjs');
const menuPositioner = require('./menu-positioner.cjs');
const menuSeparator = require('./menu-separator.cjs');
const menuTrigger = require('./menu-trigger.cjs');
const menuTriggerItem = require('./menu-trigger-item.cjs');

const Menu = Object.assign(menu.Menu, {
  Root: menu.Menu,
  Arrow: menuArrow.MenuArrow,
  ArrowTip: menuArrowTip.MenuArrowTip,
  Content: menuContent.MenuContent,
  ContextTrigger: menuContextTrigger.MenuContextTrigger,
  Item: menuItem.MenuItem,
  ItemGroup: menuItemGroup.MenuItemGroup,
  ItemGroupLabel: menuItemGroupLabel.MenuItemGroupLabel,
  OptionItem: menuOptionItem.MenuOptionItem,
  Positioner: menuPositioner.MenuPositioner,
  Separator: menuSeparator.MenuSeparator,
  Trigger: menuTrigger.MenuTrigger,
  TriggerItem: menuTriggerItem.MenuTriggerItem
});

exports.MenuArrow = menuArrow.MenuArrow;
exports.MenuArrowTip = menuArrowTip.MenuArrowTip;
exports.MenuContent = menuContent.MenuContent;
exports.useMenuContext = menuContext.useMenuContext;
exports.MenuContextTrigger = menuContextTrigger.MenuContextTrigger;
exports.MenuItem = menuItem.MenuItem;
exports.MenuItemGroup = menuItemGroup.MenuItemGroup;
exports.MenuItemGroupLabel = menuItemGroupLabel.MenuItemGroupLabel;
exports.MenuOptionItem = menuOptionItem.MenuOptionItem;
exports.MenuPositioner = menuPositioner.MenuPositioner;
exports.MenuSeparator = menuSeparator.MenuSeparator;
exports.MenuTrigger = menuTrigger.MenuTrigger;
exports.MenuTriggerItem = menuTriggerItem.MenuTriggerItem;
exports.Menu = Menu;
