// @ark-ui/react/menu/menu-context.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const createContext = require('../create-context.cjs');

const [MenuProvider, useMenuContext] = createContext.createContext({
  name: "MenuContext",
  hookName: "useMenuContext",
  providerName: "<MenuProvider />",
  strict: false
});
const [MenuMachineProvider, useMenuMachineContext] = createContext.createContext({
  name: "MenuMachineContext",
  hookName: "useMenuMachineContext",
  providerName: "<MenuMachineProvider />",
  strict: false
});
const [MenuTriggerItemProvider, useMenuTriggerItemContext] = createContext.createContext({
  name: "MenuMachineContext",
  hookName: "useMenuMachineContext",
  providerName: "<MenuMachineProvider />",
  strict: false
});

exports.MenuMachineProvider = MenuMachineProvider;
exports.MenuProvider = MenuProvider;
exports.MenuTriggerItemProvider = MenuTriggerItemProvider;
exports.useMenuContext = useMenuContext;
exports.useMenuMachineContext = useMenuMachineContext;
exports.useMenuTriggerItemContext = useMenuTriggerItemContext;
