// @ark-ui/react/menu/use-menu.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const menu = require('@zag-js/menu');
const react$1 = require('@zag-js/react');
const react = require('react');
require('../environment/index.cjs');
const useEvent = require('../use-event.cjs');
const environmentContext = require('../environment/environment-context.cjs');

function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const menu__namespace = /*#__PURE__*/_interopNamespaceDefault(menu);

const useMenu = (props = {}) => {
  const context = {
    id: react.useId(),
    getRootNode: environmentContext.useEnvironmentContext(),
    ...props,
    onOpenChange: useEvent.useEvent(props.onOpenChange),
    onSelect: useEvent.useEvent(props.onSelect),
    onValueChange: useEvent.useEvent(props.onValueChange, { sync: true })
  };
  const [state, send, machine] = react$1.useMachine(menu__namespace.machine(context), { context });
  const api = menu__namespace.connect(state, send, react$1.normalizeProps);
  return { api, machine };
};

exports.useMenu = useMenu;
