// @ark-ui/react/segment-group/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const segmentGroup = require('./segment-group.cjs');
const segmentGroupContext = require('./segment-group-context.cjs');
const segmentGroupIndicator = require('./segment-group-indicator.cjs');
const segmentGroupItem = require('./segment-group-item.cjs');
const segmentGroupItemContext = require('./segment-group-item-context.cjs');
const segmentGroupItemControl = require('./segment-group-item-control.cjs');
const segmentGroupItemText = require('./segment-group-item-text.cjs');
const segmentGroupLabel = require('./segment-group-label.cjs');

const SegmentGroup = Object.assign(segmentGroup.SegmentGroup, {
  Root: segmentGroup.SegmentGroup,
  Indicator: segmentGroupIndicator.SegmentGroupIndicator,
  Item: segmentGroupItem.SegmentGroupItem,
  ItemControl: segmentGroupItemControl.SegmentGroupItemControl,
  ItemText: segmentGroupItemText.SegmentGroupItemText,
  Label: segmentGroupLabel.SegmentGroupLabel
});

exports.useSegmentGroupContext = segmentGroupContext.useSegmentGroupContext;
exports.SegmentGroupIndicator = segmentGroupIndicator.SegmentGroupIndicator;
exports.SegmentGroupItem = segmentGroupItem.SegmentGroupItem;
exports.useSegmentGroupItemContext = segmentGroupItemContext.useSegmentGroupItemContext;
exports.SegmentGroupItemControl = segmentGroupItemControl.SegmentGroupItemControl;
exports.SegmentGroupItemText = segmentGroupItemText.SegmentGroupItemText;
exports.SegmentGroupLabel = segmentGroupLabel.SegmentGroupLabel;
exports.SegmentGroup = SegmentGroup;
