// @ark-ui/react/number-input/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const numberInput = require('./number-input.cjs');
const numberInputContext = require('./number-input-context.cjs');
const numberInputControl = require('./number-input-control.cjs');
const numberInputDecrementTrigger = require('./number-input-decrement-trigger.cjs');
const numberInputIncrementTrigger = require('./number-input-increment-trigger.cjs');
const numberInputInput = require('./number-input-input.cjs');
const numberInputLabel = require('./number-input-label.cjs');
const numberInputScrubber = require('./number-input-scrubber.cjs');

const NumberInput = Object.assign(numberInput.NumberInput, {
  Root: numberInput.NumberInput,
  Control: numberInputControl.NumberInputControl,
  DecrementTrigger: numberInputDecrementTrigger.NumberInputDecrementTrigger,
  Input: numberInputInput.NumberInputInput,
  IncrementTrigger: numberInputIncrementTrigger.NumberInputIncrementTrigger,
  Label: numberInputLabel.NumberInputLabel,
  Scrubber: numberInputScrubber.NumberInputScrubber
});

exports.useNumberInputContext = numberInputContext.useNumberInputContext;
exports.NumberInputControl = numberInputControl.NumberInputControl;
exports.NumberInputDecrementTrigger = numberInputDecrementTrigger.NumberInputDecrementTrigger;
exports.NumberInputIncrementTrigger = numberInputIncrementTrigger.NumberInputIncrementTrigger;
exports.NumberInputInput = numberInputInput.NumberInputInput;
exports.NumberInputLabel = numberInputLabel.NumberInputLabel;
exports.NumberInputScrubber = numberInputScrubber.NumberInputScrubber;
exports.NumberInput = NumberInput;
