// @ark-ui/react/number-input/index.mjs
'use client';
import { NumberInput as NumberInput$1 } from './number-input.mjs';
export { useNumberInputContext } from './number-input-context.mjs';
import { NumberInputControl } from './number-input-control.mjs';
import { NumberInputDecrementTrigger } from './number-input-decrement-trigger.mjs';
import { NumberInputIncrementTrigger } from './number-input-increment-trigger.mjs';
import { NumberInputInput } from './number-input-input.mjs';
import { NumberInputLabel } from './number-input-label.mjs';
import { NumberInputScrubber } from './number-input-scrubber.mjs';

const NumberInput = Object.assign(NumberInput$1, {
  Root: NumberInput$1,
  Control: NumberInputControl,
  DecrementTrigger: NumberInputDecrementTrigger,
  Input: NumberInputInput,
  IncrementTrigger: NumberInputIncrementTrigger,
  Label: NumberInputLabel,
  Scrubber: NumberInputScrubber
});

export { NumberInput, NumberInputControl, NumberInputDecrementTrigger, NumberInputIncrementTrigger, NumberInputInput, NumberInputLabel, NumberInputScrubber };
