// @ark-ui/anatomy/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const anatomy = require('@zag-js/anatomy');
const accordion = require('@zag-js/accordion');
const avatar = require('@zag-js/avatar');
const index = require('./carousel/index.cjs');
const checkbox = require('@zag-js/checkbox');
const index$1 = require('./color-picker/index.cjs');
const combobox = require('@zag-js/combobox');
const index$2 = require('./date-picker/index.cjs');
const dialog = require('@zag-js/dialog');
const editable = require('@zag-js/editable');
const fileUpload = require('@zag-js/file-upload');
const hoverCard = require('@zag-js/hover-card');
const menu = require('@zag-js/menu');
const numberInput = require('@zag-js/number-input');
const pagination = require('@zag-js/pagination');
const pinInput = require('@zag-js/pin-input');
const popover = require('@zag-js/popover');
const progress = require('@zag-js/progress');
const radioGroup = require('@zag-js/radio-group');
const ratingGroup = require('@zag-js/rating-group');
const index$3 = require('./segment-group/index.cjs');
const index$4 = require('./select/index.cjs');
const slider = require('@zag-js/slider');
const splitter = require('@zag-js/splitter');
const _switch = require('@zag-js/switch');
const tabs = require('@zag-js/tabs');
const tagsInput = require('@zag-js/tags-input');
const toast = require('@zag-js/toast');
const toggleGroup = require('@zag-js/toggle-group');
const tooltip = require('@zag-js/tooltip');



Object.defineProperty(exports, "accordionAnatomy", {
	enumerable: true,
	get: () => accordion.anatomy
});
Object.defineProperty(exports, "avatarAnatomy", {
	enumerable: true,
	get: () => avatar.anatomy
});
exports.carouselAnatomy = index.anatomy;
Object.defineProperty(exports, "checkboxAnatomy", {
	enumerable: true,
	get: () => checkbox.anatomy
});
exports.colorPickerAnatomy = index$1.anatomy;
Object.defineProperty(exports, "comboboxAnatomy", {
	enumerable: true,
	get: () => combobox.anatomy
});
exports.datePickerAnatomy = index$2.anatomy;
Object.defineProperty(exports, "dialogAnatomy", {
	enumerable: true,
	get: () => dialog.anatomy
});
Object.defineProperty(exports, "editableAnatomy", {
	enumerable: true,
	get: () => editable.anatomy
});
Object.defineProperty(exports, "fileUploadAnatomy", {
	enumerable: true,
	get: () => fileUpload.anatomy
});
Object.defineProperty(exports, "hoverCardAnatomy", {
	enumerable: true,
	get: () => hoverCard.anatomy
});
Object.defineProperty(exports, "menuAnatomy", {
	enumerable: true,
	get: () => menu.anatomy
});
Object.defineProperty(exports, "numberInputAnatomy", {
	enumerable: true,
	get: () => numberInput.anatomy
});
Object.defineProperty(exports, "paginationAnatomy", {
	enumerable: true,
	get: () => pagination.anatomy
});
Object.defineProperty(exports, "pinInputAnatomy", {
	enumerable: true,
	get: () => pinInput.anatomy
});
Object.defineProperty(exports, "popoverAnatomy", {
	enumerable: true,
	get: () => popover.anatomy
});
Object.defineProperty(exports, "progressAnatomy", {
	enumerable: true,
	get: () => progress.anatomy
});
Object.defineProperty(exports, "radioGroupAnatomy", {
	enumerable: true,
	get: () => radioGroup.anatomy
});
Object.defineProperty(exports, "ratingGroupAnatomy", {
	enumerable: true,
	get: () => ratingGroup.anatomy
});
exports.segmentGroupAnatomy = index$3.anatomy;
exports.selectAnatomy = index$4.anatomy;
Object.defineProperty(exports, "sliderAnatomy", {
	enumerable: true,
	get: () => slider.anatomy
});
Object.defineProperty(exports, "splitterAnatomy", {
	enumerable: true,
	get: () => splitter.anatomy
});
Object.defineProperty(exports, "switchAnatomy", {
	enumerable: true,
	get: () => _switch.anatomy
});
Object.defineProperty(exports, "tabsAnatomy", {
	enumerable: true,
	get: () => tabs.anatomy
});
Object.defineProperty(exports, "tagsInputAnatomy", {
	enumerable: true,
	get: () => tagsInput.anatomy
});
Object.defineProperty(exports, "toastAnatomy", {
	enumerable: true,
	get: () => toast.anatomy
});
Object.defineProperty(exports, "toggleGroupAnatomy", {
	enumerable: true,
	get: () => toggleGroup.anatomy
});
Object.defineProperty(exports, "tooltipAnatomy", {
	enumerable: true,
	get: () => tooltip.anatomy
});
Object.keys(anatomy).forEach(k => {
	if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
		enumerable: true,
		get: () => anatomy[k]
	});
});
