// @ark-ui/react/splitter/use-splitter.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const react$1 = require('@zag-js/react');
const splitter = require('@zag-js/splitter');
const react = require('react');
require('../environment/index.cjs');
const useEvent = require('../use-event.cjs');
const environmentContext = require('../environment/environment-context.cjs');

function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const splitter__namespace = /*#__PURE__*/_interopNamespaceDefault(splitter);

const useSplitter = (props) => {
  const initialContext = {
    id: react.useId(),
    getRootNode: environmentContext.useEnvironmentContext(),
    ...props,
    size: props.defaultSize
  };
  const context = {
    ...initialContext,
    size: props.size,
    onSizeChange: useEvent.useEvent(props.onSizeChange, { sync: true }),
    onSizeChangeEnd: useEvent.useEvent(props.onSizeChangeEnd),
    onSizeChangeStart: useEvent.useEvent(props.onSizeChangeStart)
  };
  const [state, send] = react$1.useMachine(splitter__namespace.machine(initialContext), { context });
  return splitter__namespace.connect(state, send, react$1.normalizeProps);
};

exports.useSplitter = useSplitter;
