// @zag-js/toggle-group/dist/index.mjs
// src/toggle-group.anatomy.ts
import { createAnatomy } from "@zag-js/anatomy";
var anatomy = createAnatomy("toggle-group").parts("root", "item");
var parts = anatomy.build();

// src/toggle-group.connect.ts
import { getEventKey } from "@zag-js/dom-event";
import { dataAttr, isSafari, isSelfEvent } from "@zag-js/dom-query";

// src/toggle-group.dom.ts
import { createScope, nextById, prevById, queryAll } from "@zag-js/dom-query";
import { first, last } from "@zag-js/utils";
var dom = createScope({
  getRootId: (ctx) => ctx.ids?.root ?? `toggle-group:${ctx.id}`,
  getItemId: (ctx, value) => ctx.ids?.toggle?.(value) ?? `toggle-group:${ctx.id}:${value}`,
  getRootEl: (ctx) => dom.getById(ctx, dom.getRootId(ctx)),
  getElements: (ctx) => {
    const ownerId = CSS.escape(dom.getRootId(ctx));
    const selector = `[data-ownedby='${ownerId}']:not([data-disabled])`;
    return queryAll(dom.getRootEl(ctx), selector);
  },
  getFirstEl: (ctx) => first(dom.getElements(ctx)),
  getLastEl: (ctx) => last(dom.getElements(ctx)),
  getNextEl: (ctx, id) => nextById(dom.getElements(ctx), id, ctx.currentLoop),
  getPrevEl: (ctx, id) => prevById(dom.getElements(ctx), id, ctx.currentLoop)
});

// src/toggle-group.connect.ts
function connect(state, send, normalize) {
  const value = state.context.value;
  const disabled = state.context.disabled;
  const isSingle = !state.context.multiple;
  const rovingFocus = state.context.rovingFocus;
  const isHorizontal = state.context.orientation === "horizontal";
  function getItemState(props) {
    const isDisabled = !!props.disabled || !!disabled;
    const isPressed = value?.includes(props.value);
    const id = dom.getItemId(state.context, props.value);
    const isFocused = state.context.focusedId === id;
    return {
      id,
      isDisabled,
      isPressed,
      isFocused
    };
  }
  return {
    value,
    setValue(value2) {
      send({ type: "VALUE.SET", value: value2 });
    },
    getItemState,
    rootProps: normalize.element({
      ...parts.root.attrs,
      id: dom.getRootId(state.context),
      dir: state.context.dir,
      role: isSingle ? "radiogroup" : "group",
      tabIndex: state.context.isTabbingBackward ? -1 : 0,
      "data-disabled": dataAttr(disabled),
      "data-orientation": state.context.orientation,
      "data-focus": dataAttr(state.context.focusedId != null),
      style: { outline: "none" },
      onMouseDown() {
        if (disabled)
          return;
        send("ROOT.MOUSE_DOWN");
      },
      onFocus(event) {
        if (disabled)
          return;
        if (!isSelfEvent(event) || !!state.context.isClickFocus || state.context.isTabbingBackward)
          return;
        send("ROOT.FOCUS");
      },
      onBlur() {
        if (disabled)
          return;
        send("ROOT.BLUR");
      }
    }),
    getItemProps(props) {
      const itemState = getItemState(props);
      const rovingTabIndex = itemState.isFocused ? 0 : -1;
      return normalize.button({
        ...parts.item.attrs,
        id: itemState.id,
        type: "button",
        "data-ownedby": dom.getRootId(state.context),
        "data-focus": dataAttr(itemState.isFocused),
        disabled: itemState.isDisabled,
        tabIndex: rovingFocus ? rovingTabIndex : void 0,
        // radio
        role: isSingle ? "radio" : void 0,
        "aria-checked": isSingle ? itemState.isPressed : void 0,
        "aria-pressed": isSingle ? void 0 : itemState.isPressed,
        //
        "data-disabled": dataAttr(itemState.isDisabled),
        "data-orientation": state.context.orientation,
        dir: state.context.dir,
        "data-state": itemState.isPressed ? "on" : "off",
        onFocus() {
          if (itemState.isDisabled)
            return;
          send({ type: "TOGGLE.FOCUS", id: itemState.id });
        },
        onClick(event) {
          if (itemState.isDisabled)
            return;
          send({ type: "TOGGLE.CLICK", id: itemState.id, value: props.value });
          if (isSafari()) {
            event.currentTarget.focus({ preventScroll: true });
          }
        },
        onKeyDown(event) {
          if (itemState.isDisabled || !isSelfEvent(event))
            return;
          const keyMap = {
            Tab(event2) {
              const isShiftTab = event2.shiftKey;
              send({ type: "TOGGLE.SHIFT_TAB", isShiftTab });
            },
            ArrowLeft() {
              if (!rovingFocus || !isHorizontal)
                return;
              send("TOGGLE.FOCUS_PREV");
            },
            ArrowRight() {
              if (!rovingFocus || !isHorizontal)
                return;
              send("TOGGLE.FOCUS_NEXT");
            },
            ArrowUp() {
              if (!rovingFocus || isHorizontal)
                return;
              send("TOGGLE.FOCUS_PREV");
            },
            ArrowDown() {
              if (!rovingFocus || isHorizontal)
                return;
              send("TOGGLE.FOCUS_NEXT");
            },
            Home() {
              if (!rovingFocus)
                return;
              send("TOGGLE.FOCUS_FIRST");
            },
            End() {
              if (!rovingFocus)
                return;
              send("TOGGLE.FOCUS_LAST");
            }
          };
          const exec = keyMap[getEventKey(event)];
          if (exec) {
            exec(event);
            if (event.key !== "Tab")
              event.preventDefault();
          }
        }
      });
    }
  };
}

// src/toggle-group.machine.ts
import { createMachine, guards } from "@zag-js/core";
import { raf } from "@zag-js/dom-query";
import { add, compact, isEqual, remove } from "@zag-js/utils";
var { not, and } = guards;
function machine(userContext) {
  const ctx = compact(userContext);
  return createMachine(
    {
      id: "toggle-group",
      initial: "idle",
      context: {
        value: [],
        disabled: false,
        orientation: "horizontal",
        rovingFocus: true,
        loop: true,
        ...ctx,
        focusedId: null,
        isTabbingBackward: false,
        hasFocusableToggle: false,
        isClickFocus: false,
        isWithinToolbar: false
      },
      computed: {
        currentLoop: (ctx2) => ctx2.loop && !ctx2.isWithinToolbar
      },
      entry: ["checkFocusableToggles", "checkIfWithinToolbar"],
      on: {
        "VALUE.SET": {
          actions: ["setValue"]
        },
        "TOGGLE.CLICK": {
          actions: ["setValue"]
        },
        "ROOT.MOUSE_DOWN": {
          actions: ["setClickFocus"]
        }
      },
      states: {
        idle: {
          on: {
            "ROOT.FOCUS": {
              target: "focused",
              guard: not(and("isClickFocus", "isTabbingBackward")),
              actions: ["focusFirstToggle", "clearClickFocus"]
            },
            "TOGGLE.FOCUS": {
              target: "focused",
              actions: ["setFocusedId"]
            }
          }
        },
        focused: {
          on: {
            "ROOT.BLUR": {
              target: "idle",
              actions: ["clearIsTabbingBackward"]
            },
            "TOGGLE.FOCUS": {
              actions: ["setFocusedId"]
            },
            "TOGGLE.FOCUS_NEXT": {
              actions: ["focusNextToggle"]
            },
            "TOGGLE.FOCUS_PREV": {
              actions: ["focusPrevToggle"]
            },
            "TOGGLE.FOCUS_FIRST": {
              actions: ["focusFirstToggle"]
            },
            "TOGGLE.FOCUS_LAST": {
              actions: ["focusLastToggle"]
            },
            "TOGGLE.SHIFT_TAB": {
              target: "idle",
              actions: ["setIsTabbingBackward"]
            }
          }
        }
      }
    },
    {
      guards: {
        isClickFocus: (ctx2) => ctx2.isClickFocus,
        isTabbingBackward: (ctx2) => ctx2.isTabbingBackward
      },
      actions: {
        setIsTabbingBackward(ctx2) {
          ctx2.isTabbingBackward = true;
        },
        clearIsTabbingBackward(ctx2) {
          ctx2.isTabbingBackward = false;
        },
        setClickFocus(ctx2) {
          ctx2.isClickFocus = true;
        },
        clearClickFocus(ctx2) {
          ctx2.isClickFocus = false;
        },
        checkFocusableToggles(ctx2) {
          ctx2.hasFocusableToggle = ctx2.value.length > 0;
        },
        checkIfWithinToolbar(ctx2) {
          const closestToolbar = dom.getRootEl(ctx2)?.closest("[role=toolbar]");
          ctx2.isWithinToolbar = !!closestToolbar;
        },
        setFocusedId(ctx2, evt) {
          ctx2.focusedId = evt.id;
        },
        clearFocusedId(ctx2) {
          ctx2.focusedId = null;
        },
        setValue(ctx2, evt) {
          if (!evt.value)
            return;
          let next = Array.from(ctx2.value);
          if (ctx2.multiple) {
            next = next.includes(evt.value) ? remove(next, evt.value) : add(next, evt.value);
          } else {
            next = isEqual(ctx2.value, [evt.value]) ? [] : [evt.value];
          }
          set.value(ctx2, next);
        },
        focusNextToggle(ctx2) {
          raf(() => {
            if (!ctx2.focusedId)
              return;
            dom.getNextEl(ctx2, ctx2.focusedId)?.focus({ preventScroll: true });
          });
        },
        focusPrevToggle(ctx2) {
          raf(() => {
            if (!ctx2.focusedId)
              return;
            dom.getPrevEl(ctx2, ctx2.focusedId)?.focus({ preventScroll: true });
          });
        },
        focusFirstToggle(ctx2) {
          raf(() => {
            dom.getFirstEl(ctx2)?.focus({ preventScroll: true });
          });
        },
        focusLastToggle(ctx2) {
          raf(() => {
            dom.getLastEl(ctx2)?.focus({ preventScroll: true });
          });
        }
      }
    }
  );
}
var invoke = {
  change(ctx) {
    ctx.onValueChange?.({ value: Array.from(ctx.value) });
  }
};
var set = {
  value(ctx, value) {
    if (isEqual(ctx.value, value))
      return;
    ctx.value = value;
    invoke.change(ctx);
  }
};
export {
  anatomy,
  connect,
  machine
};
