// @ark-ui/react/pagination/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const pagination = require('./pagination.cjs');
const paginationContext = require('./pagination-context.cjs');
const paginationEllipsis = require('./pagination-ellipsis.cjs');
const paginationItem = require('./pagination-item.cjs');
const paginationNextTrigger = require('./pagination-next-trigger.cjs');
const paginationPrevTrigger = require('./pagination-prev-trigger.cjs');

const Pagination = Object.assign(pagination.Pagination, {
  Root: pagination.Pagination,
  Ellipsis: paginationEllipsis.PaginationEllipsis,
  Item: paginationItem.PaginationItem,
  NextTrigger: paginationNextTrigger.PaginationNextTrigger,
  PrevTrigger: paginationPrevTrigger.PaginationPrevTrigger
});

exports.usePaginationContext = paginationContext.usePaginationContext;
exports.PaginationEllipsis = paginationEllipsis.PaginationEllipsis;
exports.PaginationItem = paginationItem.PaginationItem;
exports.PaginationNextTrigger = paginationNextTrigger.PaginationNextTrigger;
exports.PaginationPrevTrigger = paginationPrevTrigger.PaginationPrevTrigger;
exports.Pagination = Pagination;
