// @ark-ui/react/file-upload/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const fileUpload = require('./file-upload.cjs');
const fileUploadContext = require('./file-upload-context.cjs');
const fileUploadDropzone = require('./file-upload-dropzone.cjs');
const fileUploadItem = require('./file-upload-item.cjs');
const fileUploadItemDeleteTrigger = require('./file-upload-item-delete-trigger.cjs');
const fileUploadItemGroup = require('./file-upload-item-group.cjs');
const fileUploadItemName = require('./file-upload-item-name.cjs');
const fileUploadItemPreview = require('./file-upload-item-preview.cjs');
const fileUploadItemPreviewImage = require('./file-upload-item-preview-image.cjs');
const fileUploadItemSizeText = require('./file-upload-item-size-text.cjs');
const fileUploadLabel = require('./file-upload-label.cjs');
const fileUploadTrigger = require('./file-upload-trigger.cjs');

const FileUpload = Object.assign(fileUpload.FileUpload, {
  Root: fileUpload.FileUpload,
  Dropzone: fileUploadDropzone.FileUploadDropzone,
  Item: fileUploadItem.FileUploadItem,
  ItemDeleteTrigger: fileUploadItemDeleteTrigger.FileUploadItemDeleteTrigger,
  ItemGroup: fileUploadItemGroup.FileUploadItemGroup,
  ItemName: fileUploadItemName.FileUploadItemName,
  ItemPreview: fileUploadItemPreview.FileUploadItemPreview,
  ItemPreviewImage: fileUploadItemPreviewImage.FileUploadItemPreviewImage,
  ItemSizeText: fileUploadItemSizeText.FileUploadItemSizeText,
  Label: fileUploadLabel.FileUploadLabel,
  Trigger: fileUploadTrigger.FileUploadTrigger
});

exports.useFileUploadContext = fileUploadContext.useFileUploadContext;
exports.FileUploadDropzone = fileUploadDropzone.FileUploadDropzone;
exports.FileUploadItem = fileUploadItem.FileUploadItem;
exports.FileUploadItemDeleteTrigger = fileUploadItemDeleteTrigger.FileUploadItemDeleteTrigger;
exports.FileUploadItemGroup = fileUploadItemGroup.FileUploadItemGroup;
exports.FileUploadItemName = fileUploadItemName.FileUploadItemName;
exports.FileUploadItemPreviewImage = fileUploadItemPreviewImage.FileUploadItemPreviewImage;
exports.FileUploadItemSizeText = fileUploadItemSizeText.FileUploadItemSizeText;
exports.FileUploadLabel = fileUploadLabel.FileUploadLabel;
exports.FileUploadTrigger = fileUploadTrigger.FileUploadTrigger;
exports.FileUpload = FileUpload;
