// @ark-ui/react/file-upload/use-file-upload.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const fileUpload = require('@zag-js/file-upload');
const react$1 = require('@zag-js/react');
const react = require('react');
require('../environment/index.cjs');
const useEvent = require('../use-event.cjs');
const environmentContext = require('../environment/environment-context.cjs');

function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const fileUpload__namespace = /*#__PURE__*/_interopNamespaceDefault(fileUpload);

const useFileUpload = (props = {}) => {
  const initialContext = {
    id: react.useId(),
    getRootNode: environmentContext.useEnvironmentContext(),
    ...props
  };
  const context = {
    ...initialContext,
    onFileAccept: useEvent.useEvent(props.onFileAccept),
    onFileReject: useEvent.useEvent(props.onFileReject),
    onFilesChange: useEvent.useEvent(props.onFilesChange, { sync: true })
  };
  const [state, send] = react$1.useMachine(fileUpload__namespace.machine(initialContext), { context });
  return fileUpload__namespace.connect(state, send, react$1.normalizeProps);
};

exports.useFileUpload = useFileUpload;
