// @ark-ui/react/slider/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const slider = require('./slider.cjs');
const sliderContext = require('./slider-context.cjs');
const sliderControl = require('./slider-control.cjs');
const sliderLabel = require('./slider-label.cjs');
const sliderMarker = require('./slider-marker.cjs');
const sliderMarkerGroup = require('./slider-marker-group.cjs');
const sliderRange = require('./slider-range.cjs');
const sliderThumb = require('./slider-thumb.cjs');
const sliderTrack = require('./slider-track.cjs');
const sliderValueText = require('./slider-value-text.cjs');

const Slider = Object.assign(slider.Slider, {
  Root: slider.Slider,
  Control: sliderControl.SliderControl,
  Label: sliderLabel.SliderLabel,
  Marker: sliderMarker.SliderMarker,
  MarkerGroup: sliderMarkerGroup.SliderMarkerGroup,
  Range: sliderRange.SliderRange,
  Thumb: sliderThumb.SliderThumb,
  Track: sliderTrack.SliderTrack,
  ValueText: sliderValueText.SliderValueText
});

exports.useSliderContext = sliderContext.useSliderContext;
exports.SliderControl = sliderControl.SliderControl;
exports.SliderLabel = sliderLabel.SliderLabel;
exports.SliderMarker = sliderMarker.SliderMarker;
exports.SliderMarkerGroup = sliderMarkerGroup.SliderMarkerGroup;
exports.SliderRange = sliderRange.SliderRange;
exports.SliderThumb = sliderThumb.SliderThumb;
exports.SliderTrack = sliderTrack.SliderTrack;
exports.SliderValueText = sliderValueText.SliderValueText;
exports.Slider = Slider;
