// @ark-ui/react/slider/use-slider.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const react$1 = require('@zag-js/react');
const Slider = require('@zag-js/slider');
const react = require('react');
require('../environment/index.cjs');
const useEvent = require('../use-event.cjs');
const environmentContext = require('../environment/environment-context.cjs');

function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const Slider__namespace = /*#__PURE__*/_interopNamespaceDefault(Slider);

const useSlider = (props) => {
  const initialContext = {
    id: react.useId(),
    getRootNode: environmentContext.useEnvironmentContext(),
    ...props,
    value: props.defaultValue
  };
  const context = {
    ...initialContext,
    value: props.value,
    onValueChange: useEvent.useEvent(props.onValueChange, { sync: true }),
    onValueChangeEnd: useEvent.useEvent(props.onValueChangeEnd),
    onFocusChange: useEvent.useEvent(props.onFocusChange)
  };
  const [state, send] = react$1.useMachine(Slider__namespace.machine(initialContext), {
    context
  });
  return Slider__namespace.connect(state, send, react$1.normalizeProps);
};

exports.useSlider = useSlider;
