// @ark-ui/react/date-picker/use-date-picker.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const datePicker = require('@zag-js/date-picker');
const react$1 = require('@zag-js/react');
const react = require('react');
require('../environment/index.cjs');
const useEvent = require('../use-event.cjs');
const environmentContext = require('../environment/environment-context.cjs');

function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const datePicker__namespace = /*#__PURE__*/_interopNamespaceDefault(datePicker);

const useDatePicker = (props = {}) => {
  const initialContext = {
    id: react.useId(),
    getRootNode: environmentContext.useEnvironmentContext(),
    ...props,
    focusedValue: props.focusedValue ? datePicker__namespace.parse(props.focusedValue) : void 0,
    value: props.defaultValue ? datePicker__namespace.parse(props.defaultValue) : void 0
  };
  const context = {
    ...initialContext,
    value: props.value ? datePicker__namespace.parse(props.value) : void 0,
    onValueChange: useEvent.useEvent(props.onValueChange, { sync: true }),
    onFocusChange: useEvent.useEvent(props.onFocusChange),
    onViewChange: useEvent.useEvent(props.onViewChange),
    onOpenChange: useEvent.useEvent(props.onOpenChange)
  };
  const [state, send] = react$1.useMachine(datePicker__namespace.machine(initialContext), { context });
  return datePicker__namespace.connect(state, send, react$1.normalizeProps);
};

exports.useDatePicker = useDatePicker;
