// @ark-ui/react/color-picker/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const colorPicker = require('./color-picker.cjs');
const colorPickerArea = require('./color-picker-area.cjs');
const colorPickerAreaBackground = require('./color-picker-area-background.cjs');
const colorPickerAreaContext = require('./color-picker-area-context.cjs');
const colorPickerAreaThumb = require('./color-picker-area-thumb.cjs');
const colorPickerChannelInput = require('./color-picker-channel-input.cjs');
const colorPickerChannelSlider = require('./color-picker-channel-slider.cjs');
const colorPickerChannelSliderContext = require('./color-picker-channel-slider-context.cjs');
const colorPickerChannelSliderThumb = require('./color-picker-channel-slider-thumb.cjs');
const colorPickerChannelSliderTrack = require('./color-picker-channel-slider-track.cjs');
const colorPickerContent = require('./color-picker-content.cjs');
const colorPickerContext = require('./color-picker-context.cjs');
const colorPickerControl = require('./color-picker-control.cjs');
const colorPickerEyeDropperTrigger = require('./color-picker-eye-dropper-trigger.cjs');
const colorPickerFormatSelect = require('./color-picker-format-select.cjs');
const colorPickerFormatTrigger = require('./color-picker-format-trigger.cjs');
const colorPickerLabel = require('./color-picker-label.cjs');
const colorPickerPositioner = require('./color-picker-positioner.cjs');
const colorPickerSwatch = require('./color-picker-swatch.cjs');
const colorPickerSwatchGroup = require('./color-picker-swatch-group.cjs');
const colorPickerSwatchIndicator = require('./color-picker-swatch-indicator.cjs');
const colorPickerSwatchTrigger = require('./color-picker-swatch-trigger.cjs');
const colorPickerTransparencyGrid = require('./color-picker-transparency-grid.cjs');
const colorPickerTrigger = require('./color-picker-trigger.cjs');
const colorPickerValueText = require('./color-picker-value-text.cjs');
const colorPickerView = require('./color-picker-view.cjs');

const ColorPicker = Object.assign(colorPicker.ColorPicker, {
  Root: colorPicker.ColorPicker,
  Area: colorPickerArea.ColorPickerArea,
  AreaBackground: colorPickerAreaBackground.ColorPickerAreaBackground,
  AreaThumb: colorPickerAreaThumb.ColorPickerAreaThumb,
  ChannelInput: colorPickerChannelInput.ColorPickerChannelInput,
  ChannelSlider: colorPickerChannelSlider.ColorPickerChannelSlider,
  ChannelSliderThumb: colorPickerChannelSliderThumb.ColorPickerChannelSliderThumb,
  ChannelSliderTrack: colorPickerChannelSliderTrack.ColorPickerChannelSliderTrack,
  Content: colorPickerContent.ColorPickerContent,
  Control: colorPickerControl.ColorPickerControl,
  EyeDropperTrigger: colorPickerEyeDropperTrigger.ColorPickerEyeDropperTrigger,
  FormatTrigger: colorPickerFormatTrigger.ColorPickerFormatTrigger,
  FormatSelect: colorPickerFormatSelect.ColorPickerFormatSelect,
  Label: colorPickerLabel.ColorPickerLabel,
  Positioner: colorPickerPositioner.ColorPickerPositioner,
  Swatch: colorPickerSwatch.ColorPickerSwatch,
  SwatchGroup: colorPickerSwatchGroup.ColorPickerSwatchGroup,
  SwatchIndicator: colorPickerSwatchIndicator.ColorPickerSwatchIndicator,
  SwatchTrigger: colorPickerSwatchTrigger.ColorPickerSwatchTrigger,
  TransparencyGrid: colorPickerTransparencyGrid.ColorPickerTransparencyGrid,
  Trigger: colorPickerTrigger.ColorPickerTrigger,
  ValueText: colorPickerValueText.ColorPickerValueText,
  View: colorPickerView.ColorPickerView
});

exports.ColorPickerArea = colorPickerArea.ColorPickerArea;
exports.ColorPickerAreaBackground = colorPickerAreaBackground.ColorPickerAreaBackground;
exports.useColorPickerAreaContext = colorPickerAreaContext.useColorPickerAreaContext;
exports.ColorPickerAreaThumb = colorPickerAreaThumb.ColorPickerAreaThumb;
exports.ColorPickerChannelInput = colorPickerChannelInput.ColorPickerChannelInput;
exports.ColorPickerChannelSlider = colorPickerChannelSlider.ColorPickerChannelSlider;
exports.useColorPickerChannelSliderContext = colorPickerChannelSliderContext.useColorPickerChannelSliderContext;
exports.ColorPickerChannelSliderThumb = colorPickerChannelSliderThumb.ColorPickerChannelSliderThumb;
exports.ColorPickerChannelSliderTrack = colorPickerChannelSliderTrack.ColorPickerChannelSliderTrack;
exports.ColorPickerContent = colorPickerContent.ColorPickerContent;
exports.useColorPickerContext = colorPickerContext.useColorPickerContext;
exports.ColorPickerControl = colorPickerControl.ColorPickerControl;
exports.ColorPickerEyeDropperTrigger = colorPickerEyeDropperTrigger.ColorPickerEyeDropperTrigger;
exports.ColorPickerFormatSelect = colorPickerFormatSelect.ColorPickerFormatSelect;
exports.ColorPickerFormatTrigger = colorPickerFormatTrigger.ColorPickerFormatTrigger;
exports.ColorPickerLabel = colorPickerLabel.ColorPickerLabel;
exports.ColorPickerPositioner = colorPickerPositioner.ColorPickerPositioner;
exports.ColorPickerSwatch = colorPickerSwatch.ColorPickerSwatch;
exports.ColorPickerSwatchGroup = colorPickerSwatchGroup.ColorPickerSwatchGroup;
exports.ColorPickerSwatchIndicator = colorPickerSwatchIndicator.ColorPickerSwatchIndicator;
exports.ColorPickerSwatchTrigger = colorPickerSwatchTrigger.ColorPickerSwatchTrigger;
exports.ColorPickerTransparencyGrid = colorPickerTransparencyGrid.ColorPickerTransparencyGrid;
exports.ColorPickerTrigger = colorPickerTrigger.ColorPickerTrigger;
exports.ColorPickerValueText = colorPickerValueText.ColorPickerValueText;
exports.ColorPickerView = colorPickerView.ColorPickerView;
exports.ColorPicker = ColorPicker;
