// @ark-ui/react/carousel/index.mjs
'use client';
import { Carousel as Carousel$1 } from './carousel.mjs';
export { useCarouselContext } from './carousel-context.mjs';
import { CarouselControl } from './carousel-control.mjs';
import { CarouselIndicator } from './carousel-indicator.mjs';
import { CarouselIndicatorGroup } from './carousel-indicator-group.mjs';
import { CarouselItem } from './carousel-item.mjs';
import { CarouselItemGroup } from './carousel-item-group.mjs';
import { CarouselNextTrigger } from './carousel-next-trigger.mjs';
import { CarouselPrevTrigger } from './carousel-prev-trigger.mjs';
import { CarouselViewport } from './carousel-viewport.mjs';

const Carousel = Object.assign(Carousel$1, {
  Root: Carousel$1,
  Control: CarouselControl,
  Indicator: CarouselIndicator,
  IndicatorGroup: CarouselIndicatorGroup,
  NextTrigger: CarouselNextTrigger,
  PrevTrigger: CarouselPrevTrigger,
  Item: CarouselItem,
  ItemGroup: CarouselItemGroup,
  Viewport: CarouselViewport
});

export { Carousel, CarouselControl, CarouselIndicator, CarouselIndicatorGroup, CarouselItem, CarouselItemGroup, CarouselNextTrigger, CarouselPrevTrigger, CarouselViewport };
