// @ark-ui/react/carousel/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const carousel = require('./carousel.cjs');
const carouselContext = require('./carousel-context.cjs');
const carouselControl = require('./carousel-control.cjs');
const carouselIndicator = require('./carousel-indicator.cjs');
const carouselIndicatorGroup = require('./carousel-indicator-group.cjs');
const carouselItem = require('./carousel-item.cjs');
const carouselItemGroup = require('./carousel-item-group.cjs');
const carouselNextTrigger = require('./carousel-next-trigger.cjs');
const carouselPrevTrigger = require('./carousel-prev-trigger.cjs');
const carouselViewport = require('./carousel-viewport.cjs');

const Carousel = Object.assign(carousel.Carousel, {
  Root: carousel.Carousel,
  Control: carouselControl.CarouselControl,
  Indicator: carouselIndicator.CarouselIndicator,
  IndicatorGroup: carouselIndicatorGroup.CarouselIndicatorGroup,
  NextTrigger: carouselNextTrigger.CarouselNextTrigger,
  PrevTrigger: carouselPrevTrigger.CarouselPrevTrigger,
  Item: carouselItem.CarouselItem,
  ItemGroup: carouselItemGroup.CarouselItemGroup,
  Viewport: carouselViewport.CarouselViewport
});

exports.useCarouselContext = carouselContext.useCarouselContext;
exports.CarouselControl = carouselControl.CarouselControl;
exports.CarouselIndicator = carouselIndicator.CarouselIndicator;
exports.CarouselIndicatorGroup = carouselIndicatorGroup.CarouselIndicatorGroup;
exports.CarouselItem = carouselItem.CarouselItem;
exports.CarouselItemGroup = carouselItemGroup.CarouselItemGroup;
exports.CarouselNextTrigger = carouselNextTrigger.CarouselNextTrigger;
exports.CarouselPrevTrigger = carouselPrevTrigger.CarouselPrevTrigger;
exports.CarouselViewport = carouselViewport.CarouselViewport;
exports.Carousel = Carousel;
