// @ark-ui/react/accordion/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const accordion = require('./accordion.cjs');
const accordionContext = require('./accordion-context.cjs');
const accordionItem = require('./accordion-item.cjs');
const accordionItemContent = require('./accordion-item-content.cjs');
const accordionItemContext = require('./accordion-item-context.cjs');
const accordionItemIndicator = require('./accordion-item-indicator.cjs');
const accordionItemTrigger = require('./accordion-item-trigger.cjs');

const Accordion = Object.assign(accordion.Accordion, {
  Root: accordion.Accordion,
  Item: accordionItem.AccordionItem,
  ItemContent: accordionItemContent.AccordionItemContent,
  ItemIndicator: accordionItemIndicator.AccordionItemIndicator,
  ItemTrigger: accordionItemTrigger.AccordionItemTrigger
});

exports.useAccordionContext = accordionContext.useAccordionContext;
exports.AccordionItem = accordionItem.AccordionItem;
exports.AccordionItemContent = accordionItemContent.AccordionItemContent;
exports.useAccordionItemContext = accordionItemContext.useAccordionItemContext;
exports.AccordionItemIndicator = accordionItemIndicator.AccordionItemIndicator;
exports.AccordionItemTrigger = accordionItemTrigger.AccordionItemTrigger;
exports.Accordion = Accordion;
