// @ark-ui/react/rating-group/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const ratingGroup = require('./rating-group.cjs');
const ratingGroupContext = require('./rating-group-context.cjs');
const ratingGroupControl = require('./rating-group-control.cjs');
const ratingGroupItem = require('./rating-group-item.cjs');
const ratingGroupItemContext = require('./rating-group-item-context.cjs');
const ratingGroupLabel = require('./rating-group-label.cjs');

const RatingGroup = Object.assign(ratingGroup.RatingGroup, {
  Root: ratingGroup.RatingGroup,
  Control: ratingGroupControl.RatingGroupControl,
  Label: ratingGroupLabel.RatingGroupLabel,
  Item: ratingGroupItem.RatingGroupItem
});

exports.useRatingGroupContext = ratingGroupContext.useRatingGroupContext;
exports.RatingGroupControl = ratingGroupControl.RatingGroupControl;
exports.RatingGroupItem = ratingGroupItem.RatingGroupItem;
exports.useRatingGroupItemContext = ratingGroupItemContext.useRatingGroupItemContext;
exports.RatingGroupLabel = ratingGroupLabel.RatingGroupLabel;
exports.RatingGroup = RatingGroup;
