// @ark-ui/react/hover-card/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const hoverCard = require('./hover-card.cjs');
const hoverCardArrow = require('./hover-card-arrow.cjs');
const hoverCardArrowTip = require('./hover-card-arrow-tip.cjs');
const hoverCardContent = require('./hover-card-content.cjs');
const hoverCardContext = require('./hover-card-context.cjs');
const hoverCardPositioner = require('./hover-card-positioner.cjs');
const hoverCardTrigger = require('./hover-card-trigger.cjs');

const HoverCard = Object.assign(hoverCard.HoverCard, {
  Root: hoverCard.HoverCard,
  Arrow: hoverCardArrow.HoverCardArrow,
  ArrowTip: hoverCardArrowTip.HoverCardArrowTip,
  Content: hoverCardContent.HoverCardContent,
  Positioner: hoverCardPositioner.HoverCardPositioner,
  Trigger: hoverCardTrigger.HoverCardTrigger
});

exports.HoverCardArrow = hoverCardArrow.HoverCardArrow;
exports.HoverCardArrowTip = hoverCardArrowTip.HoverCardArrowTip;
exports.HoverCardContent = hoverCardContent.HoverCardContent;
exports.useHoverCardContext = hoverCardContext.useHoverCardContext;
exports.HoverCardPositioner = hoverCardPositioner.HoverCardPositioner;
exports.HoverCardTrigger = hoverCardTrigger.HoverCardTrigger;
exports.HoverCard = HoverCard;
