// @zag-js/date-picker/dist/index.mjs
// src/date-picker.anatomy.ts
import { createAnatomy } from "@zag-js/anatomy";
var anatomy = createAnatomy("date-picker").parts(
  "root",
  "label",
  "clearTrigger",
  "content",
  "control",
  "input",
  "monthSelect",
  "nextTrigger",
  "positioner",
  "prevTrigger",
  "rangeText",
  "table",
  "tableBody",
  "tableCell",
  "tableCellTrigger",
  "tableHead",
  "tableHeader",
  "tableRow",
  "trigger",
  "viewTrigger",
  "viewControl",
  "yearSelect"
);
var parts = anatomy.build();

// src/date-picker.connect.ts
import { DateFormatter as DateFormatter2, isWeekend } from "@internationalized/date";
import {
  constrainValue,
  getDayFormatter,
  getDaysInWeek,
  getDecadeRange,
  getMonthDays,
  getMonthFormatter,
  getMonthNames,
  getTodayDate,
  getUnitDuration,
  getWeekDays,
  isDateDisabled,
  isDateEqual,
  isDateInvalid,
  isDateOutsideVisibleRange,
  isDateUnavailable,
  isTodayDate,
  setMonth,
  setYear
} from "@zag-js/date-utils";
import { getEventKey, getNativeEvent } from "@zag-js/dom-event";
import { ariaAttr, dataAttr } from "@zag-js/dom-query";
import { getPlacementStyles } from "@zag-js/popper";
import { chunk } from "@zag-js/utils";

// src/date-picker.dom.ts
import { createScope, query } from "@zag-js/dom-query";
var dom = createScope({
  getRootId: (ctx) => ctx.ids?.root ?? `datepicker:${ctx.id}`,
  getTableId: (ctx, id) => ctx.ids?.table?.(id) ?? `datepicker:${ctx.id}:table:${id}`,
  getTableHeaderId: (ctx, id) => ctx.ids?.tableHeader?.(id) ?? `datepicker:${ctx.id}:thead`,
  getTableBodyId: (ctx, id) => ctx.ids?.tableBody?.(id) ?? `datepicker:${ctx.id}:tbody`,
  getTableRowId: (ctx, id) => ctx.ids?.tableRow?.(id) ?? `datepicker:${ctx.id}:tr:${id}`,
  getContentId: (ctx) => ctx.ids?.content ?? `datepicker:${ctx.id}:content`,
  getCellTriggerId: (ctx, id) => ctx.ids?.cellTrigger?.(id) ?? `datepicker:${ctx.id}:cell-trigger:${id}`,
  getPrevTriggerId: (ctx, view) => ctx.ids?.prevTrigger?.(view) ?? `datepicker:${ctx.id}:prev:${view}`,
  getNextTriggerId: (ctx, view) => ctx.ids?.nextTrigger?.(view) ?? `datepicker:${ctx.id}:next:${view}`,
  getViewTriggerId: (ctx, view) => ctx.ids?.viewTrigger?.(view) ?? `datepicker:${ctx.id}:view:${view}`,
  getClearTriggerId: (ctx) => ctx.ids?.clearTrigger ?? `datepicker:${ctx.id}:clear`,
  getControlId: (ctx) => ctx.ids?.control ?? `datepicker:${ctx.id}:control`,
  getInputId: (ctx) => ctx.ids?.input ?? `datepicker:${ctx.id}:input`,
  getTriggerId: (ctx) => ctx.ids?.trigger ?? `datepicker:${ctx.id}:trigger`,
  getPositionerId: (ctx) => ctx.ids?.positioner ?? `datepicker:${ctx.id}:positioner`,
  getMonthSelectId: (ctx) => ctx.ids?.monthSelect ?? `datepicker:${ctx.id}:month-select`,
  getYearSelectId: (ctx) => ctx.ids?.yearSelect ?? `datepicker:${ctx.id}:year-select`,
  getFocusedCell: (ctx, view = ctx.view) => query(
    dom.getContentEl(ctx),
    `[data-part=table-cell-trigger][data-view=${view}][data-focused]:not([data-outside-range])`
  ),
  getTriggerEl: (ctx) => dom.getById(ctx, dom.getTriggerId(ctx)),
  getContentEl: (ctx) => dom.getById(ctx, dom.getContentId(ctx)),
  getInputEl: (ctx) => dom.getById(ctx, dom.getInputId(ctx)),
  getYearSelectEl: (ctx) => dom.getById(ctx, dom.getYearSelectId(ctx)),
  getMonthSelectEl: (ctx) => dom.getById(ctx, dom.getMonthSelectId(ctx)),
  getClearTriggerEl: (ctx) => dom.getById(ctx, dom.getClearTriggerId(ctx)),
  getPositionerEl: (ctx) => dom.getById(ctx, dom.getPositionerId(ctx)),
  getControlEl: (ctx) => dom.getById(ctx, dom.getControlId(ctx))
});

// src/date-picker.utils.ts
import { DateFormatter } from "@internationalized/date";
import { match } from "@zag-js/utils";
function adjustStartAndEndDate(value) {
  const [startDate, endDate] = value;
  if (!startDate || !endDate)
    return value;
  return startDate.compare(endDate) <= 0 ? value : [endDate, startDate];
}
function isDateWithinRange(date, value) {
  const [startDate, endDate] = value;
  if (!startDate || !endDate)
    return false;
  return startDate.compare(date) <= 0 && endDate.compare(date) >= 0;
}
function sortDates(values) {
  return values.sort((a, b) => a.compare(b));
}
function formatValue(ctx) {
  const formatter = new DateFormatter(ctx.locale, {
    timeZone: ctx.timeZone,
    day: "2-digit",
    month: "2-digit",
    year: "numeric"
  });
  if (ctx.selectionMode === "range") {
    const [startDate, endDate] = ctx.value;
    if (!startDate || !endDate)
      return "";
    return `${formatter.format(startDate.toDate(ctx.timeZone))} - ${formatter.format(endDate.toDate(ctx.timeZone))}`;
  }
  if (ctx.selectionMode === "single") {
    const [startValue] = ctx.value;
    if (!startValue)
      return "";
    return formatter.format(startValue.toDate(ctx.timeZone));
  }
  return ctx.value.map((date) => formatter.format(date.toDate(ctx.timeZone))).join(", ");
}
function getNextTriggerLabel(view) {
  return match(view, {
    year: "Switch to next decade",
    month: "Switch to next year",
    day: "Switch to next month"
  });
}
function getPrevTriggerLabel(view) {
  return match(view, {
    year: "Switch to previous decade",
    month: "Switch to previous year",
    day: "Switch to previous month"
  });
}
function getRoleDescription(view) {
  return match(view, {
    year: "calendar decade",
    month: "calendar year",
    day: "calendar month"
  });
}
function getViewTriggerLabel(view) {
  return match(view, {
    year: "Switch to month view",
    month: "Switch to day view",
    day: "Switch to year view"
  });
}
var PLACEHOLDERS = { day: "dd", month: "mm", year: "yyyy" };
function getInputPlaceholder(locale) {
  return new DateFormatter(locale).formatToParts(/* @__PURE__ */ new Date()).map((item) => PLACEHOLDERS[item.type] ?? item.value).join("");
}
var isValidCharacter = (char, separator) => {
  if (!char)
    return true;
  return /\d/.test(char) || char === separator || char.length !== 1;
};
var ensureValidCharacters = (value, separator) => {
  return value.split("").filter((char) => isValidCharacter(char, separator)).join("");
};
function getLocaleSeparator(locale) {
  const dateFormatter = new Intl.DateTimeFormat(locale);
  const parts2 = dateFormatter.formatToParts(/* @__PURE__ */ new Date());
  const literalPart = parts2.find((part) => part.type === "literal");
  return literalPart ? literalPart.value : "/";
}

// src/date-picker.connect.ts
var pretty = (value) => value.toString().split("T")[0];
function connect(state, send, normalize) {
  const startValue = state.context.startValue;
  const endValue = state.context.endValue;
  const selectedValue = state.context.value;
  const focusedValue = state.context.focusedValue;
  const hoveredValue = state.context.hoveredValue;
  const hoveredRangeValue = hoveredValue ? adjustStartAndEndDate([selectedValue[0], hoveredValue]) : [];
  const disabled = state.context.disabled;
  const readOnly = state.context.readOnly;
  const isInteractive = state.context.isInteractive;
  const min = state.context.min;
  const max = state.context.max;
  const locale = state.context.locale;
  const timeZone = state.context.timeZone;
  const startOfWeek = state.context.startOfWeek;
  const isFocused = state.matches("focused");
  const isOpen = state.matches("open");
  const isRangePicker = state.context.selectionMode === "range";
  const isDateUnavailableFn = state.context.isDateUnavailable;
  const currentPlacement = state.context.currentPlacement;
  const popperStyles = getPlacementStyles({
    ...state.context.positioning,
    placement: currentPlacement
  });
  const separator = getLocaleSeparator(locale);
  function getMonthWeeks(from = startValue) {
    const numOfWeeks = state.context.fixedWeeks ? 6 : void 0;
    return getMonthDays(from, locale, numOfWeeks, startOfWeek);
  }
  function getMonths(props = {}) {
    const { format } = props;
    return getMonthNames(locale, format).map((label, index) => ({ label, value: index + 1 }));
  }
  function getYears() {
    return getDecadeRange(focusedValue.year).map((year) => ({
      label: year.toString(),
      value: year
    }));
  }
  function isUnavailable(date) {
    return isDateUnavailable(date, isDateUnavailableFn, locale, min, max);
  }
  function focusMonth(month) {
    const value = setMonth(focusedValue ?? getTodayDate(timeZone), month);
    send({ type: "FOCUS.SET", value });
  }
  function focusYear(year) {
    const value = setYear(focusedValue ?? getTodayDate(timeZone), year);
    send({ type: "FOCUS.SET", value });
  }
  function getYearTableCellState(props) {
    const { value, disabled: disabled2 } = props;
    const normalized = focusedValue.set({ year: value });
    const cellState = {
      isFocused: focusedValue.year === props.value,
      isSelectable: !isDateInvalid(normalized, min, max),
      isSelected: !!selectedValue.find((date) => date.year === value),
      valueText: value.toString(),
      get isDisabled() {
        return disabled2 || !cellState.isSelectable;
      }
    };
    return cellState;
  }
  function getMonthTableCellState(props) {
    const { value, disabled: disabled2 } = props;
    const normalized = focusedValue.set({ month: value });
    const formatter = getMonthFormatter(locale, timeZone);
    const cellState = {
      isFocused: focusedValue.month === props.value,
      isSelectable: !isDateInvalid(normalized, min, max),
      isSelected: !!selectedValue.find((date) => date.month === value && date.year === focusedValue.year),
      valueText: formatter.format(normalized.toDate(timeZone)),
      get isDisabled() {
        return disabled2 || !cellState.isSelectable;
      }
    };
    return cellState;
  }
  function getDayTableCellState(props) {
    const { value, disabled: disabled2, visibleRange = state.context.visibleRange } = props;
    const formatter = getDayFormatter(locale, timeZone);
    const unitDuration = getUnitDuration(state.context.visibleDuration);
    const end = visibleRange.start.add(unitDuration).subtract({ days: 1 });
    const cellState = {
      isInvalid: isDateInvalid(value, min, max),
      isDisabled: disabled2 || isDateDisabled(value, visibleRange.start, end, min, max),
      isSelected: selectedValue.some((date) => isDateEqual(value, date)),
      isUnavailable: isDateUnavailable(value, isDateUnavailableFn, locale, min, max) && !disabled2,
      isOutsideRange: isDateOutsideVisibleRange(value, visibleRange.start, end),
      isInRange: isRangePicker && (isDateWithinRange(value, selectedValue) || isDateWithinRange(value, hoveredRangeValue)),
      isFirstInRange: isRangePicker && isDateEqual(value, selectedValue[0]),
      isLastInRange: isRangePicker && isDateEqual(value, selectedValue[1]),
      isToday: isTodayDate(value, timeZone),
      isWeekend: isWeekend(value, locale),
      formattedDate: formatter.format(value.toDate(timeZone)),
      get isFocused() {
        return isDateEqual(value, focusedValue) && !cellState.isOutsideRange;
      },
      get ariaLabel() {
        if (cellState.isUnavailable)
          return `Not available. ${cellState.formattedDate}`;
        if (cellState.isSelected)
          return `Selected date. ${cellState.formattedDate}`;
        return `Choose ${cellState.formattedDate}`;
      },
      get isSelectable() {
        return !cellState.isDisabled && !cellState.isUnavailable;
      }
    };
    return cellState;
  }
  function getTableId(props) {
    const { view = "day", id } = props;
    return [view, id].filter(Boolean).join(" ");
  }
  return {
    isFocused,
    isOpen,
    view: state.context.view,
    getDaysInWeek(week, from = startValue) {
      return getDaysInWeek(week, from, locale, startOfWeek);
    },
    getOffset(duration) {
      const from = startValue.add(duration);
      return {
        visibleRange: { start: from, end: endValue.add(duration) },
        weeks: getMonthWeeks(from)
      };
    },
    getMonthWeeks,
    isUnavailable,
    weeks: getMonthWeeks(),
    weekDays: getWeekDays(getTodayDate(timeZone), startOfWeek, timeZone, locale),
    visibleRangeText: state.context.visibleRangeText,
    value: selectedValue,
    valueAsDate: selectedValue.map((date) => date.toDate(timeZone)),
    valueAsString: selectedValue.map(pretty),
    focusedValue,
    focusedValueAsDate: focusedValue?.toDate(timeZone),
    focusedValueAsString: pretty(focusedValue),
    visibleRange: state.context.visibleRange,
    selectToday() {
      const value = constrainValue(getTodayDate(timeZone), min, max);
      send({ type: "VALUE.SET", value });
    },
    setValue(values) {
      const computedValue = values.map((date) => constrainValue(date, min, max));
      send({ type: "VALUE.SET", value: computedValue });
    },
    clearValue() {
      send("VALUE.CLEAR");
    },
    setFocusedValue(value) {
      send({ type: "FOCUS.SET", value });
    },
    open() {
      send("OPEN");
    },
    close() {
      send("CLOSE");
    },
    focusMonth,
    focusYear,
    getYears,
    getMonths,
    getYearsGrid(props = {}) {
      const { columns = 1 } = props;
      return chunk(getYears(), columns);
    },
    getDecade() {
      const years = getDecadeRange(focusedValue.year);
      return { start: years.at(0), end: years.at(-1) };
    },
    getMonthsGrid(props = {}) {
      const { columns = 1, format } = props;
      return chunk(getMonths({ format }), columns);
    },
    format(value, opts = { month: "long", year: "numeric" }) {
      return new DateFormatter2(locale, opts).format(value.toDate(timeZone));
    },
    setView(view) {
      send({ type: "VIEW.SET", cell: view });
    },
    goToNext() {
      send({ type: "GOTO.NEXT", view: state.context.view });
    },
    goToPrev() {
      send({ type: "GOTO.PREV", view: state.context.view });
    },
    rootProps: normalize.element({
      ...parts.root.attrs,
      dir: state.context.dir,
      id: dom.getRootId(state.context),
      "data-state": isOpen ? "open" : "closed",
      "data-disabled": dataAttr(disabled),
      "data-readonly": dataAttr(readOnly)
    }),
    labelProps: normalize.label({
      ...parts.label.attrs,
      dir: state.context.dir,
      htmlFor: dom.getInputId(state.context),
      "data-state": isOpen ? "open" : "closed",
      "data-disabled": dataAttr(disabled),
      "data-readonly": dataAttr(readOnly)
    }),
    controlProps: normalize.element({
      ...parts.control.attrs,
      dir: state.context.dir,
      id: dom.getControlId(state.context),
      "data-disabled": dataAttr(disabled)
    }),
    rangeTextProps: normalize.element({
      ...parts.rangeText.attrs,
      dir: state.context.dir
    }),
    contentProps: normalize.element({
      ...parts.content.attrs,
      hidden: !isOpen,
      dir: state.context.dir,
      "data-state": isOpen ? "open" : "closed",
      "data-placement": currentPlacement,
      id: dom.getContentId(state.context),
      role: "application",
      "aria-roledescription": "datepicker",
      "aria-label": "calendar"
    }),
    getTableProps(props = {}) {
      const { view = "day", columns = view === "day" ? 7 : 4 } = props;
      const uid = getTableId(props);
      return normalize.element({
        ...parts.table.attrs,
        role: "grid",
        "data-columns": columns,
        "aria-roledescription": getRoleDescription(view),
        id: dom.getTableId(state.context, uid),
        "aria-readonly": ariaAttr(readOnly),
        "aria-disabled": ariaAttr(disabled),
        "aria-multiselectable": ariaAttr(state.context.selectionMode !== "single"),
        "data-view": view,
        dir: state.context.dir,
        tabIndex: -1,
        onKeyDown(event) {
          const keyMap = {
            Enter() {
              if (isUnavailable(focusedValue))
                return;
              send({ type: "TABLE.ENTER", view, columns, focus: true });
            },
            ArrowLeft() {
              send({ type: "TABLE.ARROW_LEFT", view, columns, focus: true });
            },
            ArrowRight() {
              send({ type: "TABLE.ARROW_RIGHT", view, columns, focus: true });
            },
            ArrowUp() {
              send({ type: "TABLE.ARROW_UP", view, columns, focus: true });
            },
            ArrowDown() {
              send({ type: "TABLE.ARROW_DOWN", view, columns, focus: true });
            },
            PageUp(event2) {
              send({ type: "TABLE.PAGE_UP", larger: event2.shiftKey, view, columns, focus: true });
            },
            PageDown(event2) {
              send({ type: "TABLE.PAGE_DOWN", larger: event2.shiftKey, view, columns, focus: true });
            },
            Home() {
              send({ type: "TABLE.HOME", view, columns, focus: true });
            },
            End() {
              send({ type: "TABLE.END", view, columns, focus: true });
            }
          };
          const exec = keyMap[getEventKey(event, state.context)];
          if (exec) {
            exec(event);
            event.preventDefault();
            event.stopPropagation();
          }
        },
        onPointerLeave() {
          send({ type: "TABLE.POINTER_LEAVE" });
        },
        onPointerDown() {
          send({ type: "TABLE.POINTER_DOWN", view });
        },
        onPointerUp() {
          send({ type: "TABLE.POINTER_UP", view });
        }
      });
    },
    getTableHeadProps(props = {}) {
      const { view = "day" } = props;
      return normalize.element({
        ...parts.tableHead.attrs,
        "aria-hidden": true,
        dir: state.context.dir,
        "data-view": view,
        "data-disabled": dataAttr(disabled)
      });
    },
    getTableHeaderProps(props = {}) {
      const { view = "day" } = props;
      return normalize.element({
        ...parts.tableHeader.attrs,
        dir: state.context.dir,
        "data-view": view,
        "data-disabled": dataAttr(disabled)
      });
    },
    getTableBodyProps(props = {}) {
      const { view = "day" } = props;
      return normalize.element({
        ...parts.tableBody.attrs,
        "data-view": view,
        "data-disabled": dataAttr(disabled)
      });
    },
    getTableRowProps(props = {}) {
      const { view = "day" } = props;
      return normalize.element({
        ...parts.tableRow.attrs,
        "aria-disabled": ariaAttr(disabled),
        "data-disabled": dataAttr(disabled),
        "data-view": view
      });
    },
    getDayTableCellState,
    getDayTableCellProps(props) {
      const { value } = props;
      const cellState = getDayTableCellState(props);
      return normalize.element({
        ...parts.tableCell.attrs,
        role: "gridcell",
        "aria-disabled": ariaAttr(!cellState.isSelectable),
        "aria-selected": cellState.isSelected || cellState.isInRange,
        "aria-invalid": ariaAttr(cellState.isInvalid),
        "aria-current": cellState.isToday ? "date" : void 0,
        "data-value": value.toString()
      });
    },
    getDayTableCellTriggerProps(props) {
      const { value } = props;
      const cellState = getDayTableCellState(props);
      return normalize.element({
        ...parts.tableCellTrigger.attrs,
        id: dom.getCellTriggerId(state.context, value.toString()),
        role: "button",
        dir: state.context.dir,
        tabIndex: cellState.isFocused ? 0 : -1,
        "aria-label": cellState.ariaLabel,
        "aria-disabled": ariaAttr(!cellState.isSelectable),
        "aria-invalid": ariaAttr(cellState.isInvalid),
        "data-disabled": dataAttr(!cellState.isSelectable),
        "data-selected": dataAttr(cellState.isSelected),
        "data-value": value.toString(),
        "data-view": "day",
        "data-today": dataAttr(cellState.isToday),
        "data-focused": dataAttr(cellState.isFocused),
        "data-unavailable": dataAttr(cellState.isUnavailable),
        "data-range-start": dataAttr(cellState.isFirstInRange),
        "data-range-end": dataAttr(cellState.isLastInRange),
        "data-in-range": dataAttr(cellState.isInRange),
        "data-outside-range": dataAttr(cellState.isOutsideRange),
        "data-weekend": dataAttr(cellState.isWeekend),
        onClick() {
          if (!cellState.isSelectable)
            return;
          send({ type: "CELL.CLICK", cell: "day", value });
        },
        onPointerMove(event) {
          if (event.pointerType === "touch" || !cellState.isSelectable)
            return;
          const focus = event.currentTarget.ownerDocument.activeElement !== event.currentTarget;
          send({ type: "CELL.POINTER_MOVE", cell: "day", value, focus });
        }
      });
    },
    getMonthTableCellState,
    getMonthTableCellProps(props) {
      const { value, columns } = props;
      const cellState = getMonthTableCellState(props);
      return normalize.element({
        ...parts.tableCell.attrs,
        dir: state.context.dir,
        colSpan: columns,
        role: "gridcell",
        "aria-selected": ariaAttr(cellState.isSelected),
        "data-selected": dataAttr(cellState.isSelected),
        "aria-disabled": ariaAttr(!cellState.isSelectable),
        "data-value": value
      });
    },
    getMonthTableCellTriggerProps(props) {
      const { value } = props;
      const cellState = getMonthTableCellState(props);
      return normalize.element({
        ...parts.tableCellTrigger.attrs,
        dir: state.context.dir,
        role: "button",
        id: dom.getCellTriggerId(state.context, value.toString()),
        "data-selected": dataAttr(cellState.isSelected),
        "aria-disabled": ariaAttr(!cellState.isSelectable),
        "data-disabled": dataAttr(!cellState.isSelectable),
        "data-focused": dataAttr(cellState.isFocused),
        "aria-label": cellState.valueText,
        "data-view": "month",
        "data-value": value,
        tabIndex: cellState.isFocused ? 0 : -1,
        onClick() {
          if (!cellState.isSelectable)
            return;
          send({ type: "CELL.CLICK", cell: "month", value });
        }
      });
    },
    getYearTableCellState,
    getYearTableCellProps(props) {
      const { value, columns } = props;
      const cellState = getYearTableCellState(props);
      return normalize.element({
        ...parts.tableCell.attrs,
        dir: state.context.dir,
        colSpan: columns,
        role: "gridcell",
        "aria-selected": ariaAttr(cellState.isSelected),
        "data-selected": dataAttr(cellState.isSelected),
        "aria-disabled": ariaAttr(!cellState.isSelectable),
        "data-value": value
      });
    },
    getYearTableCellTriggerProps(props) {
      const { value } = props;
      const cellState = getYearTableCellState(props);
      return normalize.element({
        ...parts.tableCellTrigger.attrs,
        dir: state.context.dir,
        role: "button",
        id: dom.getCellTriggerId(state.context, value.toString()),
        "data-selected": dataAttr(cellState.isSelected),
        "data-focused": dataAttr(cellState.isFocused),
        "aria-disabled": ariaAttr(!cellState.isSelectable),
        "data-disabled": dataAttr(!cellState.isSelectable),
        "aria-label": cellState.valueText,
        "data-value": value,
        "data-view": "year",
        tabIndex: cellState.isFocused ? 0 : -1,
        onClick() {
          if (!cellState.isSelectable)
            return;
          send({ type: "CELL.CLICK", cell: "year", value });
        }
      });
    },
    getNextTriggerProps(props = {}) {
      const { view = "day" } = props;
      return normalize.button({
        ...parts.nextTrigger.attrs,
        dir: state.context.dir,
        id: dom.getNextTriggerId(state.context, view),
        type: "button",
        "aria-label": getPrevTriggerLabel(view),
        disabled: disabled || !state.context.isNextVisibleRangeValid,
        onClick() {
          send({ type: "GOTO.NEXT", view });
        }
      });
    },
    getPrevTriggerProps(props = {}) {
      const { view = "day" } = props;
      return normalize.button({
        ...parts.prevTrigger.attrs,
        dir: state.context.dir,
        id: dom.getPrevTriggerId(state.context, view),
        type: "button",
        "aria-label": getNextTriggerLabel(view),
        disabled: disabled || !state.context.isPrevVisibleRangeValid,
        onClick() {
          send({ type: "GOTO.PREV", view });
        }
      });
    },
    clearTriggerProps: normalize.button({
      ...parts.clearTrigger.attrs,
      id: dom.getClearTriggerId(state.context),
      dir: state.context.dir,
      type: "button",
      "aria-label": "Clear dates",
      hidden: !state.context.value.length,
      onClick() {
        send("VALUE.CLEAR");
      }
    }),
    triggerProps: normalize.button({
      ...parts.trigger.attrs,
      id: dom.getTriggerId(state.context),
      dir: state.context.dir,
      type: "button",
      "data-placement": currentPlacement,
      "aria-label": isOpen ? "Close calendar" : "Open calendar",
      "aria-controls": dom.getContentId(state.context),
      "data-state": isOpen ? "open" : "closed",
      "aria-haspopup": "grid",
      disabled,
      onClick() {
        if (!isInteractive)
          return;
        send("TRIGGER.CLICK");
      }
    }),
    getViewTriggerProps(props = {}) {
      const { view = "day" } = props;
      return normalize.button({
        ...parts.viewTrigger.attrs,
        "data-view": view,
        dir: state.context.dir,
        id: dom.getViewTriggerId(state.context, view),
        type: "button",
        disabled,
        "aria-label": getViewTriggerLabel(state.context.view),
        onClick() {
          if (!isInteractive)
            return;
          send("VIEW.CHANGE");
        }
      });
    },
    getViewControlProps(props = {}) {
      const { view = "day" } = props;
      return normalize.element({
        ...parts.viewControl.attrs,
        "data-view": view,
        dir: state.context.dir
      });
    },
    inputProps: normalize.input({
      ...parts.input.attrs,
      id: dom.getInputId(state.context),
      autoComplete: "off",
      autoCorrect: "off",
      spellCheck: "false",
      dir: state.context.dir,
      name: state.context.name,
      "data-state": isOpen ? "open" : "closed",
      readOnly,
      disabled,
      placeholder: getInputPlaceholder(locale),
      defaultValue: state.context.inputValue,
      onBeforeInput(event) {
        const { data } = getNativeEvent(event);
        if (!isValidCharacter(data, separator)) {
          event.preventDefault();
        }
      },
      onFocus() {
        send("INPUT.FOCUS");
      },
      onBlur(event) {
        send({ type: "INPUT.BLUR", value: event.currentTarget.value });
      },
      onKeyDown(event) {
        if (event.key !== "Enter" || !isInteractive)
          return;
        if (isUnavailable(state.context.focusedValue))
          return;
        send({ type: "INPUT.ENTER", value: event.currentTarget.value });
      },
      onChange(event) {
        const { value } = event.target;
        send({ type: "INPUT.CHANGE", value: ensureValidCharacters(value, separator) });
      }
    }),
    monthSelectProps: normalize.select({
      ...parts.monthSelect.attrs,
      id: dom.getMonthSelectId(state.context),
      "aria-label": "Select month",
      disabled,
      dir: state.context.dir,
      defaultValue: focusedValue.month,
      onChange(event) {
        focusMonth(Number(event.currentTarget.value));
      }
    }),
    yearSelectProps: normalize.select({
      ...parts.yearSelect.attrs,
      id: dom.getYearSelectId(state.context),
      disabled,
      "aria-label": "Select year",
      dir: state.context.dir,
      defaultValue: focusedValue.year,
      onChange(event) {
        focusYear(Number(event.currentTarget.value));
      }
    }),
    positionerProps: normalize.element({
      id: dom.getPositionerId(state.context),
      ...parts.positioner.attrs,
      dir: state.context.dir,
      style: popperStyles.floating
    })
  };
}

// src/date-picker.machine.ts
import { DateFormatter as DateFormatter3 } from "@internationalized/date";
import { createMachine, guards } from "@zag-js/core";
import {
  alignDate,
  constrainValue as constrainValue2,
  formatSelectedDate,
  getAdjustedDateFn,
  getDecadeRange as getDecadeRange2,
  getEndDate,
  getNextDay,
  getNextSection,
  getPreviousDay,
  getPreviousSection,
  getTodayDate as getTodayDate2,
  isDateEqual as isDateEqual2,
  isNextVisibleRangeInvalid,
  isPreviousVisibleRangeInvalid,
  parseDateString
} from "@zag-js/date-utils";
import { trackDismissableElement } from "@zag-js/dismissable";
import { raf } from "@zag-js/dom-query";
import { createLiveRegion } from "@zag-js/live-region";
import { getPlacement } from "@zag-js/popper";
import { disableTextSelection, restoreTextSelection } from "@zag-js/text-selection";
import { compact, isEqual } from "@zag-js/utils";
var { and } = guards;
var transformContext = (ctx) => {
  const locale = ctx.locale || "en-US";
  const timeZone = ctx.timeZone || "UTC";
  const selectionMode = ctx.selectionMode || "single";
  const numOfMonths = ctx.numOfMonths || 1;
  let value = sortDates(ctx.value || []);
  value = value.map((date) => constrainValue2(date, ctx.min, ctx.max));
  let focusedValue = value[0] || ctx.focusedValue || getTodayDate2(timeZone);
  focusedValue = constrainValue2(focusedValue, ctx.min, ctx.max);
  const startValue = alignDate(focusedValue, "start", { months: numOfMonths }, locale);
  const inputValue = ctx.format?.(value) ?? formatValue({ locale, timeZone, selectionMode, value });
  return {
    locale,
    numOfMonths,
    focusedValue,
    startValue,
    timeZone,
    value,
    inputValue,
    selectionMode,
    view: "day",
    activeIndex: 0,
    hoveredValue: null,
    closeOnSelect: true,
    ...ctx,
    positioning: {
      placement: "bottom",
      ...ctx.positioning
    }
  };
};
function machine(userContext) {
  const ctx = compact(userContext);
  return createMachine(
    {
      id: "datepicker",
      initial: ctx.open ? "open" : "idle",
      context: transformContext(ctx),
      computed: {
        valueAsString: (ctx2) => ctx2.value.map((date) => formatSelectedDate(date, null, ctx2.locale, ctx2.timeZone)),
        isInteractive: (ctx2) => !ctx2.disabled && !ctx2.readOnly,
        visibleDuration: (ctx2) => ({ months: ctx2.numOfMonths }),
        endValue: (ctx2) => getEndDate(ctx2.startValue, ctx2.visibleDuration),
        visibleRange: (ctx2) => ({ start: ctx2.startValue, end: ctx2.endValue }),
        visibleRangeText(ctx2) {
          const formatter = new DateFormatter3(ctx2.locale, { month: "long", year: "numeric", timeZone: ctx2.timeZone });
          const start = formatter.format(ctx2.startValue.toDate(ctx2.timeZone));
          const end = formatter.format(ctx2.endValue.toDate(ctx2.timeZone));
          const formatted = ctx2.selectionMode === "range" ? `${start} - ${end}` : start;
          return { start, end, formatted };
        },
        isPrevVisibleRangeValid: (ctx2) => !isPreviousVisibleRangeInvalid(ctx2.startValue, ctx2.min, ctx2.max),
        isNextVisibleRangeValid: (ctx2) => !isNextVisibleRangeInvalid(ctx2.endValue, ctx2.min, ctx2.max)
      },
      activities: ["setupLiveRegion"],
      watch: {
        locale: ["setStartValue", "setInputValue"],
        focusedValue: [
          "adjustStartDate",
          "syncMonthSelectElement",
          "syncYearSelectElement",
          "focusActiveCellIfNeeded",
          "setHoveredValueIfKeyboard"
        ],
        value: ["setInputValue"],
        valueAsString: ["announceValueText"],
        inputValue: ["syncInputElement"],
        view: ["focusActiveCell"],
        open: ["toggleVisibility"]
      },
      on: {
        "VALUE.SET": {
          actions: ["setSelectedDate", "setFocusedDate"]
        },
        "VIEW.SET": {
          actions: ["setView"]
        },
        "FOCUS.SET": {
          actions: ["setFocusedDate"]
        },
        "VALUE.CLEAR": {
          actions: ["clearSelectedDate", "clearFocusedDate", "focusInputElement"]
        },
        "GOTO.NEXT": [
          { guard: "isYearView", actions: ["focusNextDecade", "announceVisibleRange"] },
          { guard: "isMonthView", actions: ["focusNextYear", "announceVisibleRange"] },
          { actions: ["focusNextPage"] }
        ],
        "GOTO.PREV": [
          { guard: "isYearView", actions: ["focusPreviousDecade", "announceVisibleRange"] },
          { guard: "isMonthView", actions: ["focusPreviousYear", "announceVisibleRange"] },
          { actions: ["focusPreviousPage"] }
        ]
      },
      states: {
        idle: {
          tags: "closed",
          on: {
            "INPUT.FOCUS": {
              target: "focused"
            },
            "TRIGGER.CLICK": {
              target: "open",
              actions: ["focusFirstSelectedDate", "focusActiveCell", "invokeOnOpen"]
            },
            OPEN: {
              target: "open",
              actions: ["invokeOnOpen"]
            }
          }
        },
        focused: {
          tags: "closed",
          on: {
            "TRIGGER.CLICK": {
              target: "open",
              actions: ["setViewToDay", "focusFirstSelectedDate", "focusActiveCell", "invokeOnOpen"]
            },
            "INPUT.CHANGE": {
              actions: ["focusParsedDate"]
            },
            "INPUT.ENTER": {
              actions: ["focusParsedDate", "selectFocusedDate"]
            },
            "INPUT.BLUR": {
              target: "idle"
            },
            "CELL.FOCUS": {
              target: "open",
              actions: ["setView", "focusActiveCell", "invokeOnOpen"]
            },
            OPEN: {
              target: "open",
              actions: ["invokeOnOpen"]
            }
          }
        },
        open: {
          tags: "open",
          activities: ["trackDismissableElement", "trackPositioning"],
          exit: ["clearHoveredDate", "resetView"],
          on: {
            "INPUT.CHANGE": {
              actions: ["focusParsedDate"]
            },
            "CELL.CLICK": [
              {
                guard: "isMonthView",
                actions: ["setFocusedMonth", "setViewToDay"]
              },
              {
                guard: "isYearView",
                actions: ["setFocusedYear", "setViewToMonth"]
              },
              {
                guard: and("isRangePicker", "hasSelectedRange"),
                actions: ["setStartIndex", "clearSelectedDate", "setFocusedDate", "setSelectedDate", "setEndIndex"]
              },
              // === Grouped transitions (based on `closeOnSelect`) ===
              {
                guard: and("isRangePicker", "isSelectingEndDate", "closeOnSelect"),
                target: "focused",
                actions: [
                  "setFocusedDate",
                  "setSelectedDate",
                  "setStartIndex",
                  "clearHoveredDate",
                  "focusInputElement",
                  "invokeOnClose"
                ]
              },
              {
                guard: and("isRangePicker", "isSelectingEndDate"),
                actions: ["setFocusedDate", "setSelectedDate", "setStartIndex", "clearHoveredDate"]
              },
              // ===
              {
                guard: "isRangePicker",
                actions: ["setFocusedDate", "setSelectedDate", "setEndIndex"]
              },
              {
                guard: "isMultiPicker",
                actions: ["setFocusedDate", "toggleSelectedDate"]
              },
              // === Grouped transitions (based on `closeOnSelect`) ===
              {
                guard: "closeOnSelect",
                target: "focused",
                actions: ["setFocusedDate", "setSelectedDate", "focusInputElement", "invokeOnClose"]
              },
              {
                actions: ["setFocusedDate", "setSelectedDate"]
              }
              // ===
            ],
            "CELL.POINTER_MOVE": {
              guard: and("isRangePicker", "isSelectingEndDate"),
              actions: ["setHoveredDate", "setFocusedDate"]
            },
            "TABLE.POINTER_LEAVE": {
              guard: "isRangePicker",
              actions: ["clearHoveredDate"]
            },
            "TABLE.POINTER_DOWN": {
              actions: ["disableTextSelection"]
            },
            "TABLE.POINTER_UP": {
              actions: ["enableTextSelection"]
            },
            "TABLE.ESCAPE": {
              target: "focused",
              actions: ["setViewToDay", "focusFirstSelectedDate", "focusTriggerElement", "invokeOnClose"]
            },
            "TABLE.ENTER": [
              {
                guard: "isMonthView",
                actions: "setViewToDay"
              },
              {
                guard: "isYearView",
                actions: "setViewToMonth"
              },
              {
                guard: and("isRangePicker", "hasSelectedRange"),
                actions: ["setStartIndex", "clearSelectedDate", "setSelectedDate", "setEndIndex"]
              },
              // === Grouped transitions (based on `closeOnSelect`) ===
              {
                guard: and("isRangePicker", "isSelectingEndDate", "closeOnSelect"),
                target: "focused",
                actions: ["setSelectedDate", "setStartIndex", "focusInputElement", "invokeOnClose"]
              },
              {
                guard: and("isRangePicker", "isSelectingEndDate"),
                actions: ["setSelectedDate", "setStartIndex"]
              },
              // ===
              {
                guard: "isRangePicker",
                actions: ["setSelectedDate", "setEndIndex", "focusNextDay"]
              },
              {
                guard: "isMultiPicker",
                actions: ["toggleSelectedDate"]
              },
              // === Grouped transitions (based on `closeOnSelect`) ===
              {
                guard: "closeOnSelect",
                target: "focused",
                actions: ["selectFocusedDate", "focusInputElement", "invokeOnClose"]
              },
              {
                actions: ["selectFocusedDate"]
              }
              // ===
            ],
            "TABLE.ARROW_RIGHT": [
              { guard: "isMonthView", actions: "focusNextMonth" },
              { guard: "isYearView", actions: "focusNextYear" },
              { actions: ["focusNextDay", "setHoveredDate"] }
            ],
            "TABLE.ARROW_LEFT": [
              { guard: "isMonthView", actions: "focusPreviousMonth" },
              { guard: "isYearView", actions: "focusPreviousYear" },
              { actions: ["focusPreviousDay"] }
            ],
            "TABLE.ARROW_UP": [
              { guard: "isMonthView", actions: "focusPreviousMonthColumn" },
              { guard: "isYearView", actions: "focusPreviousYearColumn" },
              { actions: ["focusPreviousWeek"] }
            ],
            "TABLE.ARROW_DOWN": [
              { guard: "isMonthView", actions: "focusNextMonthColumn" },
              { guard: "isYearView", actions: "focusNextYearColumn" },
              { actions: ["focusNextWeek"] }
            ],
            "TABLE.PAGE_UP": {
              actions: ["focusPreviousSection"]
            },
            "TABLE.PAGE_DOWN": {
              actions: ["focusNextSection"]
            },
            "TABLE.HOME": [
              { guard: "isMonthView", actions: ["focusFirstMonth"] },
              { guard: "isYearView", actions: ["focusFirstYear"] },
              { actions: ["focusSectionStart"] }
            ],
            "TABLE.END": [
              { guard: "isMonthView", actions: ["focusLastMonth"] },
              { guard: "isYearView", actions: ["focusLastYear"] },
              { actions: ["focusSectionEnd"] }
            ],
            "TRIGGER.CLICK": {
              target: "focused",
              actions: ["invokeOnClose"]
            },
            "VIEW.CHANGE": [
              {
                guard: "isDayView",
                actions: ["setViewToMonth"]
              },
              {
                guard: "isMonthView",
                actions: ["setViewToYear"]
              }
            ],
            DISMISS: [
              {
                guard: "isTargetFocusable",
                target: "idle",
                actions: ["setStartIndex", "invokeOnClose"]
              },
              {
                target: "focused",
                actions: ["focusTriggerElement", "setStartIndex", "invokeOnClose"]
              }
            ],
            CLOSE: {
              target: "idle",
              actions: ["setStartIndex", "invokeOnClose"]
            }
          }
        }
      }
    },
    {
      guards: {
        isDayView: (ctx2, evt) => (evt.view || ctx2.view) === "day",
        isMonthView: (ctx2, evt) => (evt.view || ctx2.view) === "month",
        isYearView: (ctx2, evt) => (evt.view || ctx2.view) === "year",
        isRangePicker: (ctx2) => ctx2.selectionMode === "range",
        hasSelectedRange: (ctx2) => ctx2.value.length === 2,
        isMultiPicker: (ctx2) => ctx2.selectionMode === "multiple",
        isTargetFocusable: (_ctx, evt) => evt.focusable,
        isSelectingEndDate: (ctx2) => ctx2.activeIndex === 1,
        closeOnSelect: (ctx2) => !!ctx2.closeOnSelect
      },
      activities: {
        trackPositioning(ctx2) {
          ctx2.currentPlacement = ctx2.positioning.placement;
          const anchorEl = dom.getControlEl(ctx2);
          const getPositionerEl = () => dom.getPositionerEl(ctx2);
          return getPlacement(anchorEl, getPositionerEl, {
            ...ctx2.positioning,
            defer: true,
            onComplete(data) {
              ctx2.currentPlacement = data.placement;
            },
            onCleanup() {
              ctx2.currentPlacement = void 0;
            }
          });
        },
        setupLiveRegion(ctx2) {
          const doc = dom.getDoc(ctx2);
          ctx2.announcer = createLiveRegion({ level: "assertive", document: doc });
          return () => ctx2.announcer?.destroy?.();
        },
        trackDismissableElement(ctx2, _evt, { send }) {
          let focusable = false;
          return trackDismissableElement(dom.getContentEl(ctx2), {
            exclude: [dom.getInputEl(ctx2), dom.getTriggerEl(ctx2), dom.getClearTriggerEl(ctx2)],
            onInteractOutside(event) {
              focusable = event.detail.focusable;
            },
            onDismiss() {
              send({ type: "DISMISS", src: "dismissable", focusable });
            },
            onEscapeKeyDown(event) {
              event.preventDefault();
              send({ type: "TABLE.ESCAPE", src: "dismissable" });
            }
          });
        }
      },
      actions: {
        setViewToDay(ctx2) {
          set.view(ctx2, "day");
        },
        setViewToMonth(ctx2) {
          set.view(ctx2, "month");
        },
        setViewToYear(ctx2) {
          set.view(ctx2, "year");
        },
        setView(ctx2, evt) {
          set.view(ctx2, evt.cell);
        },
        announceValueText(ctx2) {
          ctx2.announcer?.announce(ctx2.valueAsString.join(","), 3e3);
        },
        announceVisibleRange(ctx2) {
          const { formatted } = ctx2.visibleRangeText;
          ctx2.announcer?.announce(formatted);
        },
        disableTextSelection(ctx2) {
          disableTextSelection({ target: dom.getContentEl(ctx2), doc: dom.getDoc(ctx2) });
        },
        enableTextSelection(ctx2) {
          restoreTextSelection({ doc: dom.getDoc(ctx2), target: dom.getContentEl(ctx2) });
        },
        focusFirstSelectedDate(ctx2) {
          if (!ctx2.value.length)
            return;
          set.focusedValue(ctx2, ctx2.value[0]);
        },
        setInputValue(ctx2) {
          const input = dom.getInputEl(ctx2);
          if (!input)
            return;
          ctx2.inputValue = ctx2.format?.(ctx2.value) ?? formatValue(ctx2);
        },
        syncInputElement(ctx2) {
          const inputEl = dom.getInputEl(ctx2);
          if (!inputEl || inputEl.value === ctx2.inputValue)
            return;
          raf(() => {
            inputEl.value = ctx2.inputValue;
            inputEl.setSelectionRange(inputEl.value.length, inputEl.value.length);
          });
        },
        setFocusedDate(ctx2, evt) {
          const value = Array.isArray(evt.value) ? evt.value[0] : evt.value;
          set.focusedValue(ctx2, constrainValue2(value, ctx2.min, ctx2.max));
        },
        setFocusedMonth(ctx2, evt) {
          set.focusedValue(ctx2, ctx2.focusedValue.set({ month: evt.value }));
        },
        focusNextMonth(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.add({ months: 1 }));
        },
        focusPreviousMonth(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.subtract({ months: 1 }));
        },
        setFocusedYear(ctx2, evt) {
          set.focusedValue(ctx2, ctx2.focusedValue.set({ year: evt.value }));
        },
        setSelectedDate(ctx2, evt) {
          const values = [...ctx2.value];
          values[ctx2.activeIndex] = evt.value ?? ctx2.focusedValue;
          set.value(ctx2, adjustStartAndEndDate(values));
        },
        toggleSelectedDate(ctx2, evt) {
          const currentValue = evt.value ?? ctx2.focusedValue;
          const index = ctx2.value.findIndex((date) => isDateEqual2(date, currentValue));
          if (index === -1) {
            const values = [...ctx2.value, currentValue];
            set.value(ctx2, sortDates(values));
          } else {
            const values = [...ctx2.value];
            values.splice(index, 1);
            set.value(ctx2, sortDates(values));
          }
        },
        setHoveredDate(ctx2, evt) {
          ctx2.hoveredValue = evt.value;
        },
        clearHoveredDate(ctx2) {
          ctx2.hoveredValue = null;
        },
        selectFocusedDate(ctx2) {
          const values = [...ctx2.value];
          values[ctx2.activeIndex] = ctx2.focusedValue.copy();
          set.value(ctx2, adjustStartAndEndDate(values));
        },
        adjustStartDate(ctx2) {
          const adjust = getAdjustedDateFn(ctx2.visibleDuration, ctx2.locale, ctx2.min, ctx2.max);
          const { startDate, focusedDate } = adjust({ focusedDate: ctx2.focusedValue, startDate: ctx2.startValue });
          ctx2.startValue = startDate;
          set.focusedValue(ctx2, focusedDate);
        },
        setPreviousDate(ctx2) {
          set.focusedValue(ctx2, getPreviousDay(ctx2.focusedValue));
        },
        setNextDate(ctx2) {
          set.focusedValue(ctx2, getNextDay(ctx2.focusedValue));
        },
        focusPreviousDay(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.subtract({ days: 1 }));
        },
        focusNextDay(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.add({ days: 1 }));
        },
        focusPreviousWeek(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.subtract({ weeks: 1 }));
        },
        focusNextWeek(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.add({ weeks: 1 }));
        },
        focusNextPage(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.add(ctx2.visibleDuration));
        },
        focusPreviousPage(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.subtract(ctx2.visibleDuration));
        },
        focusSectionStart(ctx2) {
          set.focusedValue(ctx2, ctx2.startValue.copy());
        },
        focusSectionEnd(ctx2) {
          set.focusedValue(ctx2, ctx2.endValue.copy());
        },
        focusNextSection(ctx2, evt) {
          const section = getNextSection(
            ctx2.focusedValue,
            ctx2.startValue,
            evt.larger,
            ctx2.visibleDuration,
            ctx2.locale,
            ctx2.min,
            ctx2.max
          );
          if (!section)
            return;
          set.focusedValue(ctx2, section.focusedDate);
        },
        focusPreviousSection(ctx2, evt) {
          const section = getPreviousSection(
            ctx2.focusedValue,
            ctx2.startValue,
            evt.larger,
            ctx2.visibleDuration,
            ctx2.locale,
            ctx2.min,
            ctx2.max
          );
          if (!section)
            return;
          set.focusedValue(ctx2, section.focusedDate);
        },
        focusNextYear(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.add({ years: 1 }));
        },
        focusPreviousYear(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.subtract({ years: 1 }));
        },
        focusNextDecade(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.add({ years: 10 }));
        },
        focusPreviousDecade(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.subtract({ years: 10 }));
        },
        clearSelectedDate(ctx2) {
          set.value(ctx2, []);
        },
        clearFocusedDate(ctx2) {
          set.focusedValue(ctx2, getTodayDate2(ctx2.timeZone));
        },
        focusPreviousMonthColumn(ctx2, evt) {
          set.focusedValue(ctx2, ctx2.focusedValue.subtract({ months: evt.columns }));
        },
        focusNextMonthColumn(ctx2, evt) {
          set.focusedValue(ctx2, ctx2.focusedValue.add({ months: evt.columns }));
        },
        focusPreviousYearColumn(ctx2, evt) {
          set.focusedValue(ctx2, ctx2.focusedValue.subtract({ years: evt.columns }));
        },
        focusNextYearColumn(ctx2, evt) {
          set.focusedValue(ctx2, ctx2.focusedValue.add({ years: evt.columns }));
        },
        focusFirstMonth(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.set({ month: 0 }));
        },
        focusLastMonth(ctx2) {
          set.focusedValue(ctx2, ctx2.focusedValue.set({ month: 12 }));
        },
        focusFirstYear(ctx2) {
          const range = getDecadeRange2(ctx2.focusedValue.year);
          set.focusedValue(ctx2, ctx2.focusedValue.set({ year: range.at(0) }));
        },
        focusLastYear(ctx2) {
          const range = getDecadeRange2(ctx2.focusedValue.year);
          set.focusedValue(ctx2, ctx2.focusedValue.set({ year: range.at(-1) }));
        },
        setEndIndex(ctx2) {
          ctx2.activeIndex = 1;
        },
        setStartIndex(ctx2) {
          ctx2.activeIndex = 0;
        },
        focusActiveCell(ctx2) {
          raf(() => {
            dom.getFocusedCell(ctx2)?.focus({ preventScroll: true });
          });
        },
        focusActiveCellIfNeeded(ctx2, evt) {
          if (!evt.focus)
            return;
          raf(() => {
            dom.getFocusedCell(ctx2)?.focus({ preventScroll: true });
          });
        },
        setHoveredValueIfKeyboard(ctx2, evt) {
          if (!evt.type.startsWith("TABLE.ARROW") || ctx2.selectionMode !== "range" || ctx2.activeIndex === 0)
            return;
          ctx2.hoveredValue = ctx2.focusedValue.copy();
        },
        focusTriggerElement(ctx2) {
          raf(() => {
            dom.getTriggerEl(ctx2)?.focus({ preventScroll: true });
          });
        },
        focusInputElement(ctx2) {
          raf(() => {
            dom.getInputEl(ctx2)?.focus({ preventScroll: true });
          });
        },
        syncMonthSelectElement(ctx2) {
          const monthSelectEl = dom.getMonthSelectEl(ctx2);
          if (!monthSelectEl)
            return;
          monthSelectEl.value = ctx2.focusedValue.month.toString();
        },
        syncYearSelectElement(ctx2) {
          const yearSelectEl = dom.getYearSelectEl(ctx2);
          if (!yearSelectEl)
            return;
          yearSelectEl.value = ctx2.focusedValue.year.toString();
        },
        focusParsedDate(ctx2, evt) {
          ctx2.inputValue = evt.value;
          const date = parseDateString(ctx2.inputValue, ctx2.locale, ctx2.timeZone);
          set.focusedValue(ctx2, date);
        },
        resetView(ctx2, _evt, { initialContext }) {
          set.view(ctx2, initialContext.view);
        },
        setStartValue(ctx2) {
          const startValue = alignDate(ctx2.focusedValue, "start", { months: ctx2.numOfMonths }, ctx2.locale);
          ctx2.startValue = startValue;
        },
        invokeOnOpen(ctx2) {
          ctx2.onOpenChange?.({ open: true });
        },
        invokeOnClose(ctx2) {
          ctx2.onOpenChange?.({ open: false });
        },
        toggleVisibility(ctx2, _evt, { send }) {
          send({ type: ctx2.open ? "OPEN" : "CLOSE", src: "controlled" });
        }
      },
      transformContext(ctx2) {
        Object.assign(ctx2, transformContext(ctx2));
      },
      compareFns: {
        startValue: isDateEqual2,
        focusedValue: isDateEqual2,
        value: isDateEqualFn
      }
    }
  );
}
var invoke = {
  change(ctx) {
    const value = Array.from(ctx.value);
    const valueAsString = value.map((date) => date.toString());
    ctx.onValueChange?.({
      value,
      valueAsString,
      view: ctx.view
    });
  },
  focusChange(ctx) {
    const value = Array.from(ctx.value);
    const valueAsString = value.map((date) => date.toString());
    ctx.onFocusChange?.({
      focusedValue: ctx.focusedValue,
      value,
      valueAsString,
      view: ctx.view
    });
  },
  viewChange(ctx) {
    ctx.onViewChange?.({ view: ctx.view });
  }
};
var isDateEqualFn = (a, b) => {
  if (a.length !== b.length)
    return false;
  return a.every((date, index) => isDateEqual2(date, b[index]));
};
var set = {
  value(ctx, value) {
    if (isDateEqualFn(ctx.value, value))
      return;
    ctx.value = value;
    invoke.change(ctx);
  },
  focusedValue(ctx, value) {
    if (!value || isDateEqual2(ctx.focusedValue, value))
      return;
    ctx.focusedValue = value;
    invoke.focusChange(ctx);
  },
  view(ctx, value) {
    if (isEqual(ctx.view, value))
      return;
    ctx.view = value;
    invoke.viewChange(ctx);
  }
};

// src/date-picker.parse.ts
import { CalendarDate, parseDate } from "@internationalized/date";
function parse(value) {
  if (Array.isArray(value)) {
    return value.map((v) => parse(v));
  }
  if (value instanceof Date) {
    return new CalendarDate(value.getFullYear(), value.getMonth() + 1, value.getDate());
  }
  return parseDate(value);
}
export {
  anatomy,
  connect,
  machine,
  parse
};
