// @ark-ui/react/select/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const select = require('./select.cjs');
const selectClearTrigger = require('./select-clear-trigger.cjs');
const selectContent = require('./select-content.cjs');
const selectContext = require('./select-context.cjs');
const selectControl = require('./select-control.cjs');
const selectIndicator = require('./select-indicator.cjs');
const selectItem = require('./select-item.cjs');
const selectItemContext = require('./select-item-context.cjs');
const selectItemGroup = require('./select-item-group.cjs');
const selectItemGroupLabel = require('./select-item-group-label.cjs');
const selectItemIndicator = require('./select-item-indicator.cjs');
const selectItemText = require('./select-item-text.cjs');
const selectLabel = require('./select-label.cjs');
const selectPositioner = require('./select-positioner.cjs');
const selectTrigger = require('./select-trigger.cjs');
const selectValueText = require('./select-value-text.cjs');

const Select = Object.assign(select.Select, {
  Root: select.Select,
  ClearTrigger: selectClearTrigger.SelectClearTrigger,
  Content: selectContent.SelectContent,
  Control: selectControl.SelectControl,
  Indicator: selectIndicator.SelectIndicator,
  Item: selectItem.SelectItem,
  ItemGroup: selectItemGroup.SelectItemGroup,
  ItemGroupLabel: selectItemGroupLabel.SelectItemGroupLabel,
  ItemIndicator: selectItemIndicator.SelectItemIndicator,
  ItemText: selectItemText.SelectItemText,
  Label: selectLabel.SelectLabel,
  Positioner: selectPositioner.SelectPositioner,
  Trigger: selectTrigger.SelectTrigger,
  ValueText: selectValueText.SelectValueText
});

exports.SelectClearTrigger = selectClearTrigger.SelectClearTrigger;
exports.SelectContent = selectContent.SelectContent;
exports.useSelectContext = selectContext.useSelectContext;
exports.SelectControl = selectControl.SelectControl;
exports.SelectIndicator = selectIndicator.SelectIndicator;
exports.SelectItem = selectItem.SelectItem;
exports.useSelectItemContext = selectItemContext.useSelectItemContext;
exports.SelectItemGroup = selectItemGroup.SelectItemGroup;
exports.SelectItemGroupLabel = selectItemGroupLabel.SelectItemGroupLabel;
exports.SelectItemIndicator = selectItemIndicator.SelectItemIndicator;
exports.SelectItemText = selectItemText.SelectItemText;
exports.SelectLabel = selectLabel.SelectLabel;
exports.SelectPositioner = selectPositioner.SelectPositioner;
exports.SelectTrigger = selectTrigger.SelectTrigger;
exports.SelectValueText = selectValueText.SelectValueText;
exports.Select = Select;
