// @ark-ui/react/dialog/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const dialog = require('./dialog.cjs');
const dialogBackdrop = require('./dialog-backdrop.cjs');
const dialogCloseTrigger = require('./dialog-close-trigger.cjs');
const dialogContent = require('./dialog-content.cjs');
const dialogContext = require('./dialog-context.cjs');
const dialogDescription = require('./dialog-description.cjs');
const dialogPositioner = require('./dialog-positioner.cjs');
const dialogTitle = require('./dialog-title.cjs');
const dialogTrigger = require('./dialog-trigger.cjs');

const Dialog = Object.assign(dialog.Dialog, {
  Root: dialog.Dialog,
  Backdrop: dialogBackdrop.DialogBackdrop,
  CloseTrigger: dialogCloseTrigger.DialogCloseTrigger,
  Content: dialogContent.DialogContent,
  Description: dialogDescription.DialogDescription,
  Positioner: dialogPositioner.DialogPositioner,
  Title: dialogTitle.DialogTitle,
  Trigger: dialogTrigger.DialogTrigger
});

exports.DialogBackdrop = dialogBackdrop.DialogBackdrop;
exports.DialogCloseTrigger = dialogCloseTrigger.DialogCloseTrigger;
exports.DialogContent = dialogContent.DialogContent;
exports.useDialogContext = dialogContext.useDialogContext;
exports.DialogDescription = dialogDescription.DialogDescription;
exports.DialogPositioner = dialogPositioner.DialogPositioner;
exports.DialogTitle = dialogTitle.DialogTitle;
exports.DialogTrigger = dialogTrigger.DialogTrigger;
exports.Dialog = Dialog;
