// @ark-ui/react/dialog/use-dialog.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const dialog = require('@zag-js/dialog');
const react$1 = require('@zag-js/react');
const react = require('react');
require('../environment/index.cjs');
const useEvent = require('../use-event.cjs');
const environmentContext = require('../environment/environment-context.cjs');

function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const dialog__namespace = /*#__PURE__*/_interopNamespaceDefault(dialog);

const useDialog = (props = {}) => {
  const initialContext = {
    id: react.useId(),
    getRootNode: environmentContext.useEnvironmentContext(),
    ...props,
    open: props.defaultOpen ?? props.open
  };
  const context = {
    ...initialContext,
    open: props.open,
    onOpenChange: useEvent.useEvent(props.onOpenChange, { sync: true }),
    onEscapeKeyDown: useEvent.useEvent(props.onEscapeKeyDown),
    onInteractOutside: useEvent.useEvent(props.onInteractOutside)
  };
  const [state, send] = react$1.useMachine(dialog__namespace.machine(initialContext), { context });
  return dialog__namespace.connect(state, send, react$1.normalizeProps);
};

exports.useDialog = useDialog;
