// @ark-ui/react/popover/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const popover = require('./popover.cjs');
const popoverAnchor = require('./popover-anchor.cjs');
const popoverArrow = require('./popover-arrow.cjs');
const popoverArrowTip = require('./popover-arrow-tip.cjs');
const popoverCloseTrigger = require('./popover-close-trigger.cjs');
const popoverContent = require('./popover-content.cjs');
const popoverContext = require('./popover-context.cjs');
const popoverDescription = require('./popover-description.cjs');
const popoverIndicator = require('./popover-indicator.cjs');
const popoverPositioner = require('./popover-positioner.cjs');
const popoverTitle = require('./popover-title.cjs');
const popoverTrigger = require('./popover-trigger.cjs');

const Popover = Object.assign(popover.Popover, {
  Root: popover.Popover,
  Anchor: popoverAnchor.PopoverAnchor,
  Arrow: popoverArrow.PopoverArrow,
  ArrowTip: popoverArrowTip.PopoverArrowTip,
  CloseTrigger: popoverCloseTrigger.PopoverCloseTrigger,
  Content: popoverContent.PopoverContent,
  Description: popoverDescription.PopoverDescription,
  Indicator: popoverIndicator.PopoverIndicator,
  Positioner: popoverPositioner.PopoverPositioner,
  Title: popoverTitle.PopoverTitle,
  Trigger: popoverTrigger.PopoverTrigger
});

exports.PopoverAnchor = popoverAnchor.PopoverAnchor;
exports.PopoverArrow = popoverArrow.PopoverArrow;
exports.PopoverArrowTip = popoverArrowTip.PopoverArrowTip;
exports.PopoverCloseTrigger = popoverCloseTrigger.PopoverCloseTrigger;
exports.PopoverContent = popoverContent.PopoverContent;
exports.usePopoverContext = popoverContext.usePopoverContext;
exports.PopoverDescription = popoverDescription.PopoverDescription;
exports.PopoverIndicator = popoverIndicator.PopoverIndicator;
exports.PopoverPositioner = popoverPositioner.PopoverPositioner;
exports.PopoverTitle = popoverTitle.PopoverTitle;
exports.PopoverTrigger = popoverTrigger.PopoverTrigger;
exports.Popover = Popover;
