// @ark-ui/react/popover/index.mjs
'use client';
import { Popover as Popover$1 } from './popover.mjs';
import { PopoverAnchor } from './popover-anchor.mjs';
import { PopoverArrow } from './popover-arrow.mjs';
import { PopoverArrowTip } from './popover-arrow-tip.mjs';
import { PopoverCloseTrigger } from './popover-close-trigger.mjs';
import { PopoverContent } from './popover-content.mjs';
export { usePopoverContext } from './popover-context.mjs';
import { PopoverDescription } from './popover-description.mjs';
import { PopoverIndicator } from './popover-indicator.mjs';
import { PopoverPositioner } from './popover-positioner.mjs';
import { PopoverTitle } from './popover-title.mjs';
import { PopoverTrigger } from './popover-trigger.mjs';

const Popover = Object.assign(Popover$1, {
  Root: Popover$1,
  Anchor: PopoverAnchor,
  Arrow: PopoverArrow,
  ArrowTip: PopoverArrowTip,
  CloseTrigger: PopoverCloseTrigger,
  Content: PopoverContent,
  Description: PopoverDescription,
  Indicator: PopoverIndicator,
  Positioner: PopoverPositioner,
  Title: PopoverTitle,
  Trigger: PopoverTrigger
});

export { Popover, PopoverAnchor, PopoverArrow, PopoverArrowTip, PopoverCloseTrigger, PopoverContent, PopoverDescription, PopoverIndicator, PopoverPositioner, PopoverTitle, PopoverTrigger };
