// @ark-ui/react/combobox/index.mjs
'use client';
import { Combobox as Combobox$1 } from './combobox.mjs';
import { ComboboxClearTrigger } from './combobox-clear-trigger.mjs';
import { ComboboxContent } from './combobox-content.mjs';
export { useComboboxContext } from './combobox-context.mjs';
import { ComboboxControl } from './combobox-control.mjs';
import { ComboboxInput } from './combobox-input.mjs';
import { ComboboxItem } from './combobox-item.mjs';
import { ComboboxItemGroup } from './combobox-item-group.mjs';
import { ComboboxItemGroupLabel } from './combobox-item-group-label.mjs';
import { ComboboxItemIndicator } from './combobox-item-indicator.mjs';
import { ComboboxItemText } from './combobox-item-text.mjs';
import { ComboboxLabel } from './combobox-label.mjs';
import { ComboboxPositioner } from './combobox-positioner.mjs';
import { ComboboxTrigger } from './combobox-trigger.mjs';

const Combobox = Object.assign(Combobox$1, {
  Root: Combobox$1,
  ClearTrigger: ComboboxClearTrigger,
  Content: ComboboxContent,
  Control: ComboboxControl,
  Input: ComboboxInput,
  Item: ComboboxItem,
  ItemGroup: ComboboxItemGroup,
  ItemGroupLabel: ComboboxItemGroupLabel,
  ItemIndicator: ComboboxItemIndicator,
  ItemText: ComboboxItemText,
  Label: ComboboxLabel,
  Positioner: ComboboxPositioner,
  Trigger: ComboboxTrigger
});

export { Combobox, ComboboxClearTrigger, ComboboxContent, ComboboxControl, ComboboxInput, ComboboxItem, ComboboxItemGroup, ComboboxItemGroupLabel, ComboboxItemIndicator, ComboboxItemText, ComboboxLabel, ComboboxPositioner, ComboboxTrigger };
