// @ark-ui/react/combobox/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const combobox = require('./combobox.cjs');
const comboboxClearTrigger = require('./combobox-clear-trigger.cjs');
const comboboxContent = require('./combobox-content.cjs');
const comboboxContext = require('./combobox-context.cjs');
const comboboxControl = require('./combobox-control.cjs');
const comboboxInput = require('./combobox-input.cjs');
const comboboxItem = require('./combobox-item.cjs');
const comboboxItemGroup = require('./combobox-item-group.cjs');
const comboboxItemGroupLabel = require('./combobox-item-group-label.cjs');
const comboboxItemIndicator = require('./combobox-item-indicator.cjs');
const comboboxItemText = require('./combobox-item-text.cjs');
const comboboxLabel = require('./combobox-label.cjs');
const comboboxPositioner = require('./combobox-positioner.cjs');
const comboboxTrigger = require('./combobox-trigger.cjs');

const Combobox = Object.assign(combobox.Combobox, {
  Root: combobox.Combobox,
  ClearTrigger: comboboxClearTrigger.ComboboxClearTrigger,
  Content: comboboxContent.ComboboxContent,
  Control: comboboxControl.ComboboxControl,
  Input: comboboxInput.ComboboxInput,
  Item: comboboxItem.ComboboxItem,
  ItemGroup: comboboxItemGroup.ComboboxItemGroup,
  ItemGroupLabel: comboboxItemGroupLabel.ComboboxItemGroupLabel,
  ItemIndicator: comboboxItemIndicator.ComboboxItemIndicator,
  ItemText: comboboxItemText.ComboboxItemText,
  Label: comboboxLabel.ComboboxLabel,
  Positioner: comboboxPositioner.ComboboxPositioner,
  Trigger: comboboxTrigger.ComboboxTrigger
});

exports.ComboboxClearTrigger = comboboxClearTrigger.ComboboxClearTrigger;
exports.ComboboxContent = comboboxContent.ComboboxContent;
exports.useComboboxContext = comboboxContext.useComboboxContext;
exports.ComboboxControl = comboboxControl.ComboboxControl;
exports.ComboboxInput = comboboxInput.ComboboxInput;
exports.ComboboxItem = comboboxItem.ComboboxItem;
exports.ComboboxItemGroup = comboboxItemGroup.ComboboxItemGroup;
exports.ComboboxItemGroupLabel = comboboxItemGroupLabel.ComboboxItemGroupLabel;
exports.ComboboxItemIndicator = comboboxItemIndicator.ComboboxItemIndicator;
exports.ComboboxItemText = comboboxItemText.ComboboxItemText;
exports.ComboboxLabel = comboboxLabel.ComboboxLabel;
exports.ComboboxPositioner = comboboxPositioner.ComboboxPositioner;
exports.ComboboxTrigger = comboboxTrigger.ComboboxTrigger;
exports.Combobox = Combobox;
