// @ark-ui/react/combobox/use-combobox.mjs
'use client';
import * as combobox from '@zag-js/combobox';
import { useMachine, normalizeProps } from '@zag-js/react';
import { useMemo, useId } from 'react';
import { createSplitProps } from '../create-split-props.mjs';
import '../environment/index.mjs';
import { useEvent } from '../use-event.mjs';
import { useEnvironmentContext } from '../environment/environment-context.mjs';

const useCombobox = (props) => {
  const [collectionOptions, comboboxProps] = createSplitProps()(props, [
    "isItemDisabled",
    "itemToValue",
    "itemToString",
    "items"
  ]);
  const collection = useMemo(
    () => combobox.collection(collectionOptions),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    Object.values(collectionOptions)
  );
  const initialContext = {
    id: useId(),
    getRootNode: useEnvironmentContext(),
    collection,
    ...comboboxProps,
    value: props.defaultValue
  };
  const context = {
    ...initialContext,
    collection,
    value: props.value,
    onValueChange: useEvent(props.onValueChange, { sync: true }),
    onInputValueChange: useEvent(props.onInputValueChange, { sync: true }),
    onHighlightChange: useEvent(props.onHighlightChange),
    onOpenChange: useEvent(props.onOpenChange)
  };
  const [state, send] = useMachine(combobox.machine(initialContext), {
    context
  });
  return combobox.connect(state, send, normalizeProps);
};

export { useCombobox };
