// @ark-ui/react/progress/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const progress = require('./progress.cjs');
const progressCircle = require('./progress-circle.cjs');
const progressCircleRange = require('./progress-circle-range.cjs');
const progressCircleTrack = require('./progress-circle-track.cjs');
const progressContext = require('./progress-context.cjs');
const progressIndicator = require('./progress-indicator.cjs');
const progressLabel = require('./progress-label.cjs');
const progressRange = require('./progress-range.cjs');
const progressTrack = require('./progress-track.cjs');
const progressValueText = require('./progress-value-text.cjs');

const Progress = Object.assign(progress.Progress, {
  Root: progress.Progress,
  Label: progressLabel.ProgressLabel,
  Track: progressTrack.ProgressTrack,
  Range: progressRange.ProgressRange,
  ValueText: progressValueText.ProgressValueText,
  Indicator: progressIndicator.ProgressIndicator,
  Circle: progressCircle.ProgressCircle,
  CircleTrack: progressCircleTrack.ProgressCircleTrack,
  CircleRange: progressCircleRange.ProgressCircleRange
});

exports.ProgressCircle = progressCircle.ProgressCircle;
exports.ProgressCircleRange = progressCircleRange.ProgressCircleRange;
exports.ProgressCircleTrack = progressCircleTrack.ProgressCircleTrack;
exports.useProgressContext = progressContext.useProgressContext;
exports.ProgressIndicator = progressIndicator.ProgressIndicator;
exports.ProgressLabel = progressLabel.ProgressLabel;
exports.ProgressRange = progressRange.ProgressRange;
exports.ProgressTrack = progressTrack.ProgressTrack;
exports.ProgressValueText = progressValueText.ProgressValueText;
exports.Progress = Progress;
