// @ark-ui/react/progress/index.mjs
'use client';
import { Progress as Progress$1 } from './progress.mjs';
import { ProgressCircle } from './progress-circle.mjs';
import { ProgressCircleRange } from './progress-circle-range.mjs';
import { ProgressCircleTrack } from './progress-circle-track.mjs';
export { useProgressContext } from './progress-context.mjs';
import { ProgressIndicator } from './progress-indicator.mjs';
import { ProgressLabel } from './progress-label.mjs';
import { ProgressRange } from './progress-range.mjs';
import { ProgressTrack } from './progress-track.mjs';
import { ProgressValueText } from './progress-value-text.mjs';

const Progress = Object.assign(Progress$1, {
  Root: Progress$1,
  Label: ProgressLabel,
  Track: ProgressTrack,
  Range: ProgressRange,
  ValueText: ProgressValueText,
  Indicator: ProgressIndicator,
  Circle: ProgressCircle,
  CircleTrack: ProgressCircleTrack,
  CircleRange: ProgressCircleRange
});

export { Progress, ProgressCircle, ProgressCircleRange, ProgressCircleTrack, ProgressIndicator, ProgressLabel, ProgressRange, ProgressTrack, ProgressValueText };
