// @zag-js/react/dist/index.mjs
"use client"

// src/index.ts
import { mergeProps } from "@zag-js/core";

// src/normalize-props.ts
import { createNormalizer } from "@zag-js/types";
var normalizeProps = createNormalizer((v) => v);

// src/portal.tsx
import { Children, useReducer } from "react";
import { createPortal } from "react-dom";

// src/use-layout-effect.ts
import { useEffect, useLayoutEffect } from "react";
var useSafeLayoutEffect = typeof document !== "undefined" ? useLayoutEffect : useEffect;

// src/portal.tsx
import { Fragment, jsx } from "react/jsx-runtime";
var Portal = (props) => {
  const { children, container, disabled, getRootNode } = props;
  const [, forceUpdate] = useReducer((c) => c + 1, 0);
  useSafeLayoutEffect(() => {
    forceUpdate();
  }, []);
  const isServer = typeof window === "undefined";
  if (isServer || disabled)
    return /* @__PURE__ */ jsx(Fragment, { children });
  const doc = getRootNode?.().ownerDocument ?? document;
  const mountNode = container?.current ?? doc.body;
  return /* @__PURE__ */ jsx(Fragment, { children: Children.map(children, (child) => createPortal(child, mountNode)) });
};

// src/use-snapshot.ts
import { snapshot, subscribe } from "@zag-js/store";

// ../../utilities/core/src/equal.ts
var isArrayLike = (value) => value?.constructor.name === "Array";
var isEqual = (a, b) => {
  if (Object.is(a, b))
    return true;
  if (a == null && b != null || a != null && b == null)
    return false;
  if (typeof a?.isEqual === "function" && typeof b?.isEqual === "function") {
    return a.isEqual(b);
  }
  if (typeof a === "function" && typeof b === "function") {
    return a.toString() === b.toString();
  }
  if (isArrayLike(a) && isArrayLike(b)) {
    return Array.from(a).toString() === Array.from(b).toString();
  }
  if (!(typeof a === "object") || !(typeof b === "object"))
    return false;
  const keys = Object.keys(b ?? /* @__PURE__ */ Object.create(null));
  const length = keys.length;
  for (let i = 0; i < length; i++) {
    const hasKey = Reflect.has(a, keys[i]);
    if (!hasKey)
      return false;
  }
  for (let i = 0; i < length; i++) {
    const key = keys[i];
    if (!isEqual(a[key], b[key]))
      return false;
  }
  return true;
};

// ../../utilities/core/src/object.ts
function compact(obj) {
  if (!isPlainObject(obj) || obj === void 0) {
    return obj;
  }
  const keys = Reflect.ownKeys(obj).filter((key) => typeof key === "string");
  const filtered = {};
  for (const key of keys) {
    const value = obj[key];
    if (value !== void 0) {
      filtered[key] = compact(value);
    }
  }
  return filtered;
}
var isPlainObject = (value) => {
  return value && typeof value === "object" && value.constructor === Object;
};

// src/use-snapshot.ts
import { createProxy as createProxyToCompare, isChanged } from "proxy-compare";
import ReactExport, { useCallback, useEffect as useEffect3, useMemo, useRef as useRef2, useSyncExternalStore } from "react";

// src/use-update-effect.ts
import { useEffect as useEffect2, useRef } from "react";
function useUpdateEffect(callback, deps) {
  const render = useRef(false);
  const effect = useRef(false);
  useEffect2(() => {
    const mounted = render.current;
    const run = mounted && effect.current;
    if (run) {
      return callback();
    }
    effect.current = true;
  }, deps);
  useEffect2(() => {
    render.current = true;
    return () => {
      render.current = false;
    };
  }, []);
}

// src/use-snapshot.ts
var { use } = ReactExport;
var targetCache = /* @__PURE__ */ new WeakMap();
function useSnapshot(service, options) {
  const { actions, context, sync: notifyInSync } = options ?? {};
  const lastSnapshot = useRef2();
  const lastAffected = useRef2();
  const currSnapshot = useSyncExternalStore(
    useCallback((callback) => subscribe(service.state, callback, notifyInSync), [notifyInSync]),
    () => {
      const nextSnapshot = snapshot(service.state, use);
      try {
        if (lastSnapshot.current && lastAffected.current && !isChanged(lastSnapshot.current, nextSnapshot, lastAffected.current, /* @__PURE__ */ new WeakMap())) {
          return lastSnapshot.current;
        }
      } catch (e) {
      }
      return nextSnapshot;
    },
    () => snapshot(service.state, use)
  );
  service.setOptions({ actions });
  const ctx = useMemo(() => compact(context ?? {}), [context]);
  useUpdateEffect(() => {
    const entries = Object.entries(ctx);
    const equality = entries.map(([key, value]) => ({
      key,
      curr: value,
      prev: currSnapshot.context[key],
      equal: isEqual(currSnapshot.context[key], value)
    }));
    const allEqual = equality.every(({ equal }) => equal);
    if (!allEqual) {
      service.setContext(ctx);
    }
  }, [ctx]);
  const currAffected = /* @__PURE__ */ new WeakMap();
  useEffect3(() => {
    lastSnapshot.current = currSnapshot;
    lastAffected.current = currAffected;
  });
  const proxyCache = useMemo(() => /* @__PURE__ */ new WeakMap(), []);
  return createProxyToCompare(currSnapshot, currAffected, proxyCache, targetCache);
}

// src/use-actor.ts
function useActor(service) {
  const state = useSnapshot(service);
  return [state, service.send];
}

// src/use-constant.ts
import { useRef as useRef3 } from "react";
function useConstant(fn) {
  const ref = useRef3();
  if (!ref.current)
    ref.current = { v: fn() };
  return ref.current.v;
}

// src/use-service.ts
function useService(machine, options) {
  const { state: hydratedState, context } = options ?? {};
  const service = useConstant(() => {
    const instance = typeof machine === "function" ? machine() : machine;
    return context ? instance.withContext(context) : instance;
  });
  useSafeLayoutEffect(() => {
    service.start(hydratedState);
    return () => {
      service.stop();
    };
  }, []);
  return service;
}

// src/use-machine.ts
function useMachine(machine, options) {
  const service = useService(machine, options);
  const state = useSnapshot(service, options);
  return [state, service.send, service];
}
export {
  Portal,
  mergeProps,
  normalizeProps,
  useActor,
  useMachine
};
