// @ark-ui/react/editable/index.mjs
'use client';
import { Editable as Editable$1 } from './editable.mjs';
import { EditableArea } from './editable-area.mjs';
import { EditableCancelTrigger } from './editable-cancel-trigger.mjs';
export { useEditableContext } from './editable-context.mjs';
import { EditableControl } from './editable-control.mjs';
import { EditableEditTrigger } from './editable-edit-trigger.mjs';
import { EditableInput } from './editable-input.mjs';
import { EditableLabel } from './editable-label.mjs';
import { EditablePreview } from './editable-preview.mjs';
import { EditableSubmitTrigger } from './editable-submit-trigger.mjs';

const Editable = Object.assign(Editable$1, {
  Root: Editable$1,
  Area: EditableArea,
  CancelTrigger: EditableCancelTrigger,
  Control: EditableControl,
  EditTrigger: EditableEditTrigger,
  Input: EditableInput,
  Label: EditableLabel,
  Preview: EditablePreview,
  SubmitTrigger: EditableSubmitTrigger
});

export { Editable, EditableArea, EditableCancelTrigger, EditableControl, EditableEditTrigger, EditableInput, EditableLabel, EditablePreview, EditableSubmitTrigger };
