// @ark-ui/react/editable/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const editable = require('./editable.cjs');
const editableArea = require('./editable-area.cjs');
const editableCancelTrigger = require('./editable-cancel-trigger.cjs');
const editableContext = require('./editable-context.cjs');
const editableControl = require('./editable-control.cjs');
const editableEditTrigger = require('./editable-edit-trigger.cjs');
const editableInput = require('./editable-input.cjs');
const editableLabel = require('./editable-label.cjs');
const editablePreview = require('./editable-preview.cjs');
const editableSubmitTrigger = require('./editable-submit-trigger.cjs');

const Editable = Object.assign(editable.Editable, {
  Root: editable.Editable,
  Area: editableArea.EditableArea,
  CancelTrigger: editableCancelTrigger.EditableCancelTrigger,
  Control: editableControl.EditableControl,
  EditTrigger: editableEditTrigger.EditableEditTrigger,
  Input: editableInput.EditableInput,
  Label: editableLabel.EditableLabel,
  Preview: editablePreview.EditablePreview,
  SubmitTrigger: editableSubmitTrigger.EditableSubmitTrigger
});

exports.EditableArea = editableArea.EditableArea;
exports.EditableCancelTrigger = editableCancelTrigger.EditableCancelTrigger;
exports.useEditableContext = editableContext.useEditableContext;
exports.EditableControl = editableControl.EditableControl;
exports.EditableEditTrigger = editableEditTrigger.EditableEditTrigger;
exports.EditableInput = editableInput.EditableInput;
exports.EditableLabel = editableLabel.EditableLabel;
exports.EditablePreview = editablePreview.EditablePreview;
exports.EditableSubmitTrigger = editableSubmitTrigger.EditableSubmitTrigger;
exports.Editable = Editable;
