// @ark-ui/react/editable/use-editable.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const editable = require('@zag-js/editable');
const react$1 = require('@zag-js/react');
const react = require('react');
require('../environment/index.cjs');
const useEvent = require('../use-event.cjs');
const environmentContext = require('../environment/environment-context.cjs');

function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const editable__namespace = /*#__PURE__*/_interopNamespaceDefault(editable);

const useEditable = (props = {}) => {
  const initialContext = {
    id: react.useId(),
    getRootNode: environmentContext.useEnvironmentContext(),
    ...props,
    value: props.defaultValue
  };
  const context = {
    ...initialContext,
    value: props.value,
    onValueChange: useEvent.useEvent(props.onValueChange, { sync: true }),
    onEdit: useEvent.useEvent(props.onEdit),
    onValueCommit: useEvent.useEvent(props.onValueCommit),
    onValueRevert: useEvent.useEvent(props.onValueRevert)
  };
  const [state, send] = react$1.useMachine(editable__namespace.machine(initialContext), { context });
  return editable__namespace.connect(state, send, react$1.normalizeProps);
};

exports.useEditable = useEditable;
