// @ark-ui/react/tags-input/index.mjs
'use client';
import { TagsInput as TagsInput$1 } from './tags-input.mjs';
import { TagsInputClearTrigger } from './tags-input-clear-trigger.mjs';
export { useTagsInputContext } from './tags-input-context.mjs';
import { TagsInputControl } from './tags-input-control.mjs';
import { TagsInputInput } from './tags-input-input.mjs';
import { TagsInputItem } from './tags-input-item.mjs';
export { useTagsInputItemContext } from './tags-input-item-context.mjs';
import { TagsInputItemDeleteTrigger } from './tags-input-item-delete-trigger.mjs';
import { TagsInputItemInput } from './tags-input-item-input.mjs';
import { TagsInputItemText } from './tags-input-item-text.mjs';
import { TagsInputLabel } from './tags-input-label.mjs';

const TagsInput = Object.assign(TagsInput$1, {
  Root: TagsInput$1,
  ClearTrigger: TagsInputClearTrigger,
  Control: TagsInputControl,
  Input: TagsInputInput,
  Item: TagsInputItem,
  ItemDeleteTrigger: TagsInputItemDeleteTrigger,
  ItemInput: TagsInputItemInput,
  ItemText: TagsInputItemText,
  Label: TagsInputLabel
});

export { TagsInput, TagsInputClearTrigger, TagsInputControl, TagsInputInput, TagsInputItem, TagsInputItemDeleteTrigger, TagsInputItemInput, TagsInputItemText, TagsInputLabel };
