// @ark-ui/react/tags-input/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const tagsInput = require('./tags-input.cjs');
const tagsInputClearTrigger = require('./tags-input-clear-trigger.cjs');
const tagsInputContext = require('./tags-input-context.cjs');
const tagsInputControl = require('./tags-input-control.cjs');
const tagsInputInput = require('./tags-input-input.cjs');
const tagsInputItem = require('./tags-input-item.cjs');
const tagsInputItemContext = require('./tags-input-item-context.cjs');
const tagsInputItemDeleteTrigger = require('./tags-input-item-delete-trigger.cjs');
const tagsInputItemInput = require('./tags-input-item-input.cjs');
const tagsInputItemText = require('./tags-input-item-text.cjs');
const tagsInputLabel = require('./tags-input-label.cjs');

const TagsInput = Object.assign(tagsInput.TagsInput, {
  Root: tagsInput.TagsInput,
  ClearTrigger: tagsInputClearTrigger.TagsInputClearTrigger,
  Control: tagsInputControl.TagsInputControl,
  Input: tagsInputInput.TagsInputInput,
  Item: tagsInputItem.TagsInputItem,
  ItemDeleteTrigger: tagsInputItemDeleteTrigger.TagsInputItemDeleteTrigger,
  ItemInput: tagsInputItemInput.TagsInputItemInput,
  ItemText: tagsInputItemText.TagsInputItemText,
  Label: tagsInputLabel.TagsInputLabel
});

exports.TagsInputClearTrigger = tagsInputClearTrigger.TagsInputClearTrigger;
exports.useTagsInputContext = tagsInputContext.useTagsInputContext;
exports.TagsInputControl = tagsInputControl.TagsInputControl;
exports.TagsInputInput = tagsInputInput.TagsInputInput;
exports.TagsInputItem = tagsInputItem.TagsInputItem;
exports.useTagsInputItemContext = tagsInputItemContext.useTagsInputItemContext;
exports.TagsInputItemDeleteTrigger = tagsInputItemDeleteTrigger.TagsInputItemDeleteTrigger;
exports.TagsInputItemInput = tagsInputItemInput.TagsInputItemInput;
exports.TagsInputItemText = tagsInputItemText.TagsInputItemText;
exports.TagsInputLabel = tagsInputLabel.TagsInputLabel;
exports.TagsInput = TagsInput;
