// @ark-ui/react/tags-input/use-tags-input.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const react$1 = require('@zag-js/react');
const tagsInput = require('@zag-js/tags-input');
const react = require('react');
require('../environment/index.cjs');
const useEvent = require('../use-event.cjs');
const environmentContext = require('../environment/environment-context.cjs');

function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const tagsInput__namespace = /*#__PURE__*/_interopNamespaceDefault(tagsInput);

const useTagsInput = (props) => {
  const initialContext = {
    id: react.useId(),
    getRootNode: environmentContext.useEnvironmentContext(),
    ...props,
    value: props.defaultValue
  };
  const context = {
    ...initialContext,
    value: props.value,
    onValueChange: useEvent.useEvent(props.onValueChange, { sync: true }),
    onValueInvalid: useEvent.useEvent(props.onValueInvalid),
    onHighlightChange: useEvent.useEvent(props.onHighlightChange)
  };
  const [state, send] = react$1.useMachine(tagsInput__namespace.machine(initialContext), { context });
  return tagsInput__namespace.connect(state, send, react$1.normalizeProps);
};

exports.useTagsInput = useTagsInput;
