// @ark-ui/react/radio-group/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const radioGroup = require('./radio-group.cjs');
const radioGroupContext = require('./radio-group-context.cjs');
const radioGroupIndicator = require('./radio-group-indicator.cjs');
const radioGroupItem = require('./radio-group-item.cjs');
const radioGroupItemContext = require('./radio-group-item-context.cjs');
const radioGroupItemControl = require('./radio-group-item-control.cjs');
const radioGroupItemText = require('./radio-group-item-text.cjs');
const radioGroupLabel = require('./radio-group-label.cjs');

const RadioGroup = Object.assign(radioGroup.RadioGroup, {
  Root: radioGroup.RadioGroup,
  Indicator: radioGroupIndicator.RadioGroupIndicator,
  Item: radioGroupItem.RadioGroupItem,
  ItemControl: radioGroupItemControl.RadioGroupItemControl,
  ItemText: radioGroupItemText.RadioGroupItemText,
  Label: radioGroupLabel.RadioGroupLabel
});

exports.useRadioGroupContext = radioGroupContext.useRadioGroupContext;
exports.RadioGroupItem = radioGroupItem.RadioGroupItem;
exports.useRadioGroupItemContext = radioGroupItemContext.useRadioGroupItemContext;
exports.RadioGroupItemControl = radioGroupItemControl.RadioGroupItemControl;
exports.RadioGroupItemText = radioGroupItemText.RadioGroupItemText;
exports.RadioGroupLabel = radioGroupLabel.RadioGroupLabel;
exports.RadioGroup = RadioGroup;
